% Potential reviewers and e-mail addresses 
% DANIEL FALUSH: d.falush@ucc.ie
% MATTHEW HAHN: mwh@indiana.edu 
% Debra Bessen: debra_bessen@nymc.edu
% Xavier Didelot: xavier.didelot@gmail.com
% Brian Golding: Golding@McMaster.CA

% TODO:
% 1. Not yet referenced figure - \label{fig:mowgli-sim}
% 2. Not yet referenced figure - \label{fig:sim2}
% 3. Not yet referenced figure - \label{fig:sim3}
% 
% Backup of the work
% swiftgen:/usr/projects/strep/gainloss/recombination

\documentclass[12pt]{article}

\usepackage{times}
% \usepackage[T1]{fontenc}
% \usepackage{ae,aecompl}

% amsmath package, useful for mathematical formulas
\usepackage{amsmath}
% amssymb package, useful for mathematical symbols
\usepackage{amssymb}

% graphicx package, useful for including eps and pdf graphics
% include graphics with the command \includegraphics
\usepackage{graphicx}

% cite package, to clean up citations in the main text. Do not remove.
%\usepackage{cite}

\usepackage{color} 

% Use doublespacing - comment out for single spacing
\usepackage{setspace} 

% To rotate tables using sidewaystable instead of just table.
\usepackage{rotating}

\usepackage{natbib}
\usepackage{../../latex/sty/genres}

\usepackage{xr}
\externaldocument{supplementary}

% Text layout
\oddsidemargin 0in
\evensidemargin 0in
\topmargin -.5in
\textwidth 6.5in
\textheight 9in

% Bold the 'Figure #' in the caption and separate it with a period
% Captions will be left justified
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}

% FIXME: Remove the followings if you have to when submitting the manuscript.
\providecommand{\tabularnewline}{\\}
\newcommand{\lyxdot}{.}
\newcommand{\comment}[1]{}

\makeatother

\pagestyle{myheadings}
%% ** EDIT HERE **
\markright{Horizontal gene transfer in Streptococcus}

\begin{document}

\begin{titlepage}

\title{Replacing and additive horizontal \\
gene transfer in {\em Streptococcus} }

\author{Sang Chul Choi$^{1}$, Matthew D.\ Rasmussen$^{1}$, 
Melissa J.\ Hubisz$^{1}$, \\
Ilan Gronau$^{1}$,
Michael J.\ Stanhope$^{2}$, and Adam Siepel$^{1}$}

\date{ }
\maketitle

\begin{footnotesize}
\begin{center}
$^1$Department of Biological Statistics and Computational Biology,\\
Cornell University, Ithaca, NY 14853, USA
\\[1ex]
$^2$Department of Population Medicine and Diagnostic Sciences,\\
College of Veterinary Medicine, Cornell University, Ithaca, NY 14853, USA
\\
\end{center}
\end{footnotesize}

\vspace{1in}

\begin{tabular}{lp{4.5in}}
{\bf Submission type:}& Research Article
\vspace{1ex}\\
{\bf Keywords:}&Bacterial evolutionary
genomics, recombination, {\em Streptococcus pyogenes}, {\em Streptococcus
  dysgalactiae} 
\vspace{1ex}\\
{\bf Running Head:}&Horizontal gene transfer in {\em Streptococcus}
\vspace{1ex}\\ 
{\bf Corresponding Author:}&
\begin{minipage}[t]{4in}
 Adam Siepel\\
 102E Weill Hall, Cornell University\\
 Ithaca, NY 14853\\
 Phone: +1-607-254-1157\\
 Fax: +1-607-255-4698\\
 Email: acs4@cornell.edu
\end{minipage}
\end{tabular}

\thispagestyle{empty}
\end{titlepage}

\doublespacing

% MBE limit is 350 words
\section*{Abstract}

The prominent role of horizontal gene transfer (HGT) in the evolution of
bacteria is now well documented, yet few studies have differentiated
between the mechanisms of homologous recombination and additive
integration.  These processes leave distinct phylogenetic signatures,
because homologous recombination causes a gene in one lineage to be
replaced by a homolog from another lineage (``replacing HGT''), while
additive integration causes addition without replacement (``additive
HGT'').  Here we make use of these signatures in a genome-wide
investigation of HGT in the important human pathogen {\em Streptococcus
  pyogenes} (SPY) and its close relatives {\em S.\ dysgalactiae} subspecies
{\em equisimilis} (SDE) and {\em S.\ dysgalactiae} subspecies {\em
  dysgalactiae} (SDD).  Using recently developed statistical models and
computational methods, we find evidence for abundant gene flow of both
kinds within the SPY and SDE clades, and of reduced levels of exchange
between SPY and SDD.  In addition, our analysis strongly supports a
previously reported, but unconfirmed, finding of a pronounced asymmetry in
SPY-SDE gene flow, favoring the SPY-to-SDE direction, which may be
associated with increasing virulence of pathogenic SDE.  We find much
stronger evidence for SPY-SDE gene flow among replacing rather than
additive transfers, suggesting a primary influence from homologous
recombination between co-occurring SPY and SDE cells in human hosts.
Virulence genes are correlated with transfer events, but this correlation
is driven by additive, not replacing, HGTs.  The genes affected by additive
HGTs are enriched for functions having to do with transposition,
recombination, and DNA integration, while replacing HGTs appear to
influence a more diverse set of genes.  Additive transfers are also
associated with evidence of positive selection.  These findings shed new
light on the manner in which HGT has shaped pathogenic bacterial genomes
during relatively recent evolutionary time.

\thispagestyle{empty}
\clearpage

\section*{Introduction}

During the last twenty years, it has become increasingly apparent that
horizontal gene transfer (HGT) has played a major role in genomic
evolution, particularly among bacteria and other microbes
\citep{Smith1992,Smith1993,Ochman2000,Koonin2001}.  Indeed, it is likely
that HGT has been sufficiently prevalent throughout evolutionary history
that no present-day gene can trace an unbroken history of vertical descent
to a common ancestor for all species \citep{Zhaxybayeva2011}.  This would
mean that no phylogenetic marker can be used to reconstruct a
universal tree of life.

For HGT to occur in bacteria, DNA must first be transferred from a donor to
a recipient cell by transformation, transduction, or conjugation, then must
be integrated into the recipient cell's genome (setting aside the case of
replication-proficient plasmids).  This integration typically occurs in one
of two major ways: either the new sequence replaces a homologous sequence
through the process of homologous recombination (similar to ``gene
conversion'' in sexually reproducing organisms), or it is acquired through
an additive (non-replacing) integration process \citep{Thomas2005}.  These
two types of transfers leave distinct molecular evolutionary signatures and
will be referred to in this article as ``replacing'' and ``additive'' HGTs,
respectively (fig.\ \ref{fig:hgt}).  The rates at which these processes
occur depend on many factors, including the co-occurrence of bacterial
species in the environment,
%the propensity of donor cells to release intact DNA,
the degree of competence of recipient cells, sequence similarity between
the donor and recipient, and barriers to conjugation such as surface
exclusion.  In addition, the persistence of horizontally transferred DNA
segments in present-day species depends strongly on the subsequent effects
of natural selection.  Not surprisingly, rates of HGT appear to vary
considerably across groups of bacteria \citep[e.g.,][]{Feil2001}.

HGT in the {\em Streptococcus} genus is of particular interest.  This group
of Gram-positive bacteria contains the human pathogens {\em S.\ pyogenes}
(the cause of Group A streptococcal infections, including pharyngitis,
impetigo, cellulitis, necrotizing fasciitis, and rheumatic fever), {\em S.\
  pneumoniae} (bacterial pneumonia), {\em S.\ agalactiae} (neonatal sepsis),
{\em S.\ mutans} (dental caries), and {\em S.\ dysgalactiae} ssp.\ {\em
  equisimilis} (cellulitis, peritonitis, pneumonia, and other infections),
as well as the agricultural pathogens {\em S.\ dysgalactiae} ssp.\ {\em
  dysgalactiae} (mastitis in cows, ewes, and goats), {\em S.\ equi} ssp.\
{\em equi} (strangles in horses), and {\em S.\ canis} (various infections
in dogs and other animals).  Many of these species are concentrated in the
predominantly beta-hemolytic {\em Pyogenes} group.  Despite their
relatively recent evolutionary divergence, these {\em Pyogenes}-group
species have adapted to a remarkable variety of distinct ecological niches.
Accordingly, they display high rates of recombination and HGT, as well as
substantial evidence of positive selection
\citep{Feil2001,Marri2006,Anisimova2007,Lefebure2007,Suzuki2011}.  Complete
genome sequences are now available for several species in this group
\citep[e.g.,][]{Ferretti2001,Holden2009,Shimomura2011,Suzuki2011},
providing a rich resource for the study of the genomic evolution of
pathogenic bacteria over relatively recent time scales.

Within the {\em Pyogenes} group, several studies have focused on the issue
of HGT between the human-colonizing species {\em S.\ pyogenes} (here
denoted SPY) and {\em S.\ dysgalactiae} ssp.\ {\em equisimilis} (SDE).
Until recently, SDE was considered to be primarily a commensal organism
\citep{Vandamme1996}, unlike SPY, but it has become increasingly apparent
that it also has an important pathogenetic role, with a disease spectrum
similar to that of SPY \citep{Brandt2009}.  Shared virulence genes are well
documented between the two species \citep[e.g.,][]{Davies2007a}, genetic
exchange presumably having been enabled both by their close evolutionary
relationship and shared ecological niches, but
% Towers et al 2004
% Maxted and Potter, 1967
% Simpson et al., 1992
% Sriprakash et al., 1996
in a widely noted phylogenetic study of seven loci and more than 200
isolates, \cite{Kalia2001} also found extensive evidence of SPY-SDE gene
flow among housekeeping genes.  Interestingly, they observed a pronounced
bias in the direction of gene flow, favoring the SPY-to-SDE direction.
Gene flow has been confirmed in subsequent studies
\citep{Kalia2004,Davies2005,Davies2007a,Davies2007}, but the evidence for
directional asymmetry is mixed, with some indications of gene flow in the
opposite (SDE-to-SPY) direction \citep{Sachse2002}.  This issue has been
further complicated by the eventual retraction of Kalia et al.'s
\citeyearpar{Kalia2001} paper, due to the authors' inability to replicate
their findings and concerns about sample contamination.  The authors argued
that whole-genome sequencing may be required to definitively resolve open
questions about gene flow between these species.  These issues are of
particular interest because of the possibility of increasing virulence of
SDE due, at least in part, to gene flow from SPY.

Most recent studies of gene flow in bacteria have been based on relatively
simple phylogenetic methods that make use of incongruity of inferred gene
trees across loci \citep[e.g.,][]{Kalia2001,Kalia2004,Ahmad2009}.  These
methods are sensitive to errors in phylogenetic reconstruction, and ignore
information from branch lengths and from phylogenetic correlation at
adjacent loci.  Recently alternative, model-based approaches have been
proposed for the case of homologous recombination (``replacing HGTs'')
\citep{Didelot2007,Didelot2010}.  Using Bayesian principles and Markov
chain Monte Carlo (MCMC) techniques, these methods allow for uncertainty in
the phylogeny, make use of branch lengths and correlations across loci, and
allow for inference not only of individual recombination events but also of
parameters describing global rates and patterns of recombination.  In
addition, several recent methods have been introduced to detect HGT in a
phylogenetic framework, by parsimoniously reconciling reconstructed gene
trees with a given species tree, allowing for duplication, transfer, and
loss (DTL) events \citep{Merkle2010,David2011,Doyon2011,Doyon2011b}.  These
new model- and parsimony-based methods are complementary in many respects,
and could be particularly useful in combination.

In this article, we take a fresh look at the issue of HGT in the {\em
  Pyogenes} group, making use of newly available complete genome sequences
for SPY, SDE, and the related species {\em S.\ dysgalactiae} ssp.\ {\em
  dysgalactiae} (SDD), as well as new statistical and parsimony-based
methods for 
analysis.  The combined use of these methods allows us to examine both
replacing and additive HGTs, and compare their genome-wide effects.  We
find evidence for abundant gene flow within SPY, within SDE, and between
SPY and SDE, and for greatly reduced exchange between SPY and SDD.  In
addition, our genome-wide analysis supports Kalia et al.'s
\citeyearpar{Kalia2001} previously unconfirmed finding of a pronounced
preference for the SPY-to-SDE direction in SPY-SDE gene flow.
Interestingly, this property is much more evident for replacing than for
additive transfers.  We also examine correlations of gene transfer
events with functional categories and positive selection.
%find evidence for an association
%between virulence genes and additive, but not replacing, HGTs.  
Our results
have been made available as tracks in our recently released {\em
  Streptococcus} genome browser.

\section*{Methods}

\subsection*{Genome sequences and alignments}

Our primary analysis was based on five complete genome sequences, including
two representatives of {\em S.\ pyogenes} (accessions NC\_004070
[\citealp{Beres2002}; SPY1] and NC\_008024 [\citealp{Beres2006}, SPY2]),
two of {\em S.\ dysgalactiae} ssp.\ {\em equisimilis} (CP002215
[\citealp{Suzuki2011}; SDE1] and NC\_012891 [\citealp{Shimomura2011};
SDE2]), and one of {\em S.\ dysgalactiae} ssp.\ {\em dysgalactiae}
(CM001076 [\citealp{Suzuki2011}; SDD]).  In addition, we used 
\textit{S.\ equi} ssp.\textit{\ equi} strain 4047 (NC\_012471
[\citealp{Holden2009}; SEE]) as an outgroup in our parsimony-based
analysis (supplementary table \ref{tab:genome}).
 
For the model-based analysis of replacing transfers, we obtained
high-quality alignments of SPY1, SPY2, SDE1, SDE2, and SDD using the
pipeline detailed by Didelot and colleagues
\citep{Didelot2007,Didelot2010}.  Briefly, we first aligned these five
genomes using progressiveMauve v2.3.1 \citep{Darling2004,Darling2010} with
default options (supplementary fig.\ \ref{fig:mauve}), and then used
stripSubsetLCBs (from the ClonalOrigin package; \citealp{Didelot2010}) to
identify 276 blocks of sequence alignments that exceed a length threshold
of 1,500 bp.  This conservative threshold was selected to avoid biases from
boundary effects in short alignment blocks, based on a preliminary
analysis.  Two blocks containing long gaps were excluded from the analysis.
The final set of 274 alignment blocks contained 1,155,016 columns with
relatively few gaps (0.2\% of characters) and covered 1,106 of the 1,951
genes in the SPY1 genome.  Note that this alignment procedure effectively
focused on the ``core genome,'' by discarding regions that were not
conserved among the five bacterial individuals in question.

For the parsimony-based analysis, we began with 2,314 gene family clusters
representing all six genomes (including SEE), as described by
\cite{Suzuki2011}.  We aligned protein sequences corresponding to each
family using MUSCLE \citep{Edgar2004a}, and then obtained nucleotide
alignments by reverse translation, using the genomic sequences as a guide.
Next we identified likely intragenic recombinations using the Single
Breakpoint Recombination (SBP) method from the HyPhy package
\citep{KosakovskyPond2006}.  Based on the Akaike Information Criterion
(AIC), SBP identified at least one topology-altering recombination in
47.5\% (1094) of gene families. These alignments were split into putative
non-recombining gene fragments.  The final set of alignments consisted of
3,408 gene fragment families, about half (1,794) of which contained
single gene from each of the six species.  The remaining families ranged
from as few as two genes (417 fragments) to as many as 59 (2 fragments)
(supplementary fig.\ \ref{fig:famsizes}).  These alignments were
representative of the ``pan genome,'' including regions that frequently
turn over (``dispensable'' regions), rather than just the ``core genome''
\citep[e.g.,][]{Tettelin2005,Lefebure2007}.

\subsection*{Model-based analysis of replacing HGTs}

{\bf ClonalFrame and ClonalOrigin.}  To study replacing HGTs, we made use
of the ClonalOrigin program recently developed by \cite{Didelot2010}.
Given a set of alignment blocks, ClonalOrigin samples from the posterior
distribution of {\em recombinant trees} using a Markov chain Monte Carlo
(MCMC) algorithm.  A recombinant tree consists of a pre-estimated rooted
phylogeny with branch lengths (a ``clonal frame''), augmented by a set of
{\em recombinant edges}. Each recombinant edge is defined by the two points
along the branches of the clonal frame that correspond to the {\em donor}
and {\em recipient} in the hypothesized recombination event, and by the
start and end of the affected genomic interval.  The donor can pre-date the
recipient, to allow for a delayed coalescence, but it cannot post-date it
(supplementary fig.\ \ref{fig:clonalorigin}).
Another program, called ClonalFrame
\citep{Didelot2007}, can be used to infer the phylogeny that serves as the
basis of the recombinant trees.  ClonalFrame makes use of a similar, but
somewhat simpler, probabilistic model.

{\bf Fitting the model.}  Following \cite{Didelot2010}, we used ClonalFrame
(v1.1) to estimate a clonal frame for our five-way alignments, running the
algorithm for $10^4$ burn-in iterations followed by another $10^4$ sampling
iterations.  A series of seven additional sampling runs with random
initializations converged to essentially identical estimates, indicating
stable convergence.  We then used an initial run of ClonalOrigin
(subversion r19) to estimate global parameters including the mutation rate,
recombination rate, and average recombinant tract length. We applied the method
separately to each of the 274 blocks, in all cases using $10^6$ burn-in
iterations, $10^7$ sampling iterations, and sub-sampling every $10^5$
iterations. A global estimate for each parameter was then obtained by
taking the block-length weighted median of the block-specific posterior
means, which down-weighed less reliable estimates obtained from shorter
alignment blocks.

{\bf Sampling recombinant trees}.  We ran ClonalOrigin a second time on all
alignment blocks, this time fixing the clonal frame and global model
parameters at their pre-estimated values, and collected samples of
recombinant trees across the genome.  For each block, we ran the sampler
for $10^7$ burn-in iterations and $10^8$ sampling iterations, sub-sampling
every $10^5$ iterations, resulting in 1,001 recombinant trees per alignment
block.  We performed a replicate of the two stages of the ClonalOrigin
inference to ensure adequate convergence of the sampler.

% A few alignment blocks were not successfully finished even after a month of
% computation. Because those were not expected to affect the global estimates of
% population parameters significantly, we used only finished blocks to estimate
% mutation rate, recombination rate, and average tract length.

{\bf Summary statistics.}  We used various summary statistics to describe
the sampled recombinant trees.
First, we recorded the relative frequencies of the 105 possible
rooted trees at each site in the alignments.  Second, we recorded the
frequencies of sampled recombinant edges, grouping them by the associated
donor and recipient branches in the clonal frame \citep{Didelot2010}, and
took ratios of these sampled counts with corresponding expected values
under the assumed 
prior distribution.  Third, we recorded the frequencies of these edges
types per site.  Finally, we defined a general, branch-independent {\em
  recombination intensity} at each site by summing the posterior
probabilities of all recombination edges.  In some cases, we also computed
intensities for particular types of replacing transfers, such as those
between the SPY and SDE clades, or those producing topologies different
from the clonal frame
(see \textbf{Supplementary Material}).

{\bf Significance testing}.  We used simulations to assess the significance
of the
prior-normalized counts of recombinant edges.  First,
we generated 100 replicate data sets under the prior model, with the same
block number and blocks lengths as the real data.  We then applied our
entire inference procedure to these data sets, and computed the same ratios
of posterior estimates to prior expectations as for the real data.  We then
compared the ratios estimated from real data with this empirical null
distribution.  Empirical one-sided $p$-values were computed as the fraction
of null estimates greater or less than the values estimated from real data.
This approach had the advantage of not only assessing the significance of
departures from prior expectations, but also correcting for any systematic
biases imposed by ClonalOrigin
(see \textbf{Supplementary Material}).

To compare the rates of SPY-to-SDE and SDE-to-SPY gene flow, we computed,
for each of the 1,001 genome-wide collections of recombinant trees, the
number of recombinant edges from any of the three SPY branches (SPY1, SPY2,
and SPY) to any of the three SDE branches (SDE1, SDE2, and SDE), and the
number of recombinant edges in the reverse direction.  We then took the
ratio of these two counts.  This gave us 1,001 samples from the posterior
distribution of the ratio of SPY-to-SDE and SDE-to-SPY recombinant edges,
from which we could estimate a posterior mean (2.0) and 95\% credible
interval (1.8--2.2).  None of the sampled values was as small as the prior
estimate of the ratio (1.4), indicating strong support in the data for a
preference for the SPY-to-SDE direction.

\subsection*{Parsimony-based analysis}

% move stuff down about gene tree estimation

{\bf Mowgli.} To study additive HGTs, we used a parsimony-based method
for reconciling gene and species trees called Mowgli \citep{Doyon2011}.
Given a gene tree and a species tree, Mowgli finds a reconciliation scenario 
that minimizes the number of gene duplication, loss, and
transfer (DLT) events, considering tree topologies with branching orders
(``labeled histories''; \citealp{Edwards1970}).  We estimated
unrooted gene trees for the nucleotide alignments for each of our
putative non-recombining gene fragment using RAxML \citep{Stamatakis2006},
then applied Mowgli to each of these trees together with the species
tree estimated by ClonalFrame (augmented with the SEE outgroup).
Since Mowgli required rooted gene trees, we ran Mowgli for all possible
rootings of each gene tree, and chose the scenario that produced the
minimum DLT score. In the case of multiple most parsimonious
reconciliations, one was chosen uniformly at random.

{\bf Distinguishing between replacing and additive HGTs.}  The transfer
events inferred by Mowgli had to be classified as replacing or additive in
a post-processing step because the program did not distinguish between
these types of events.  A naive approach would be to infer a replacing HGT
whenever an inferred transfer coincided exactly with an inferred loss in
the recipient lineage.  However, we found that loss events were difficult
to place accurately because they were frequently pushed toward the root of
the tree due to reconstruction error or parallel losses in multiple
lineages.  Therefore, we classified a transfer event as additive whenever a
descendant of the recipient lineage in the species tree contained both a
gene descended from the transferred copy and a gene not descended from that
copy.  All other transfers were considered replacing HGTs.  This approach
was conservative about calling additive HGTs, and labeled any HGT that
could plausibly be explained by replacing transfers as such (see
supplementary fig.\ \ref{fig:calling-transfers}).

{\bf Statistical enrichment}. As with replacing HGTs, we classified
transfer events by donor and recipient branch, and recorded the number of
inferred additive HGTs of each type.  We compared these numbers to
expectations based on simulations that assumed constant rates of
duplication, loss, and transfer across the species tree, given the total
numbers of events inferred across all 2,314 gene families divided by the
total branch lengths of the trees (see \textbf{Supplementary Material})
% (supplementary fig.\ \ref{fig:mowgli-sim}).

%Ilan: add some detail about simulations, or point to supp

\subsection*{Gene category associations and virulence genes}

We assigned genes to Gene Ontology (GO) categories by comparing the {\em
  Streptococcus} genes to bacterial proteins from the Uniref90 database
using BLASTP, and then assigning the same GO classification as the target
gene of the uniProt GOA database if the match had an $E$-value of
$<1.0\times10^{-5}.$  Virulence genes were identified using the methods
described by \cite{Suzuki2011}. A gene family was assigned a given
classification if any of its genes was assigned that classification.  To
test for associations with replacing gene transfers we used the
recombination intensity estimated for each gene, and for additive gene
transfers we used the number of additive transfers inferred for each
family.  To test for significance, we performed a Mann-Whitney $U$ test of
the values (recombination intensities or numbers of transfers) associated
with a given category vs.\ the values for the all other genes/families.  We
used the \cite{Benjamini1995} method to correct for multiple comparisons.

\section*{Results}

\subsection*{Clonal frame and global parameter estimates}
% The 5-species tree was
% (((1:0.045557,2:0.045557)6:0.195330,3:0.240887)8:0.089874,(4:0.074544,5:0.074544)7:0.256218)9:0.000000
By applying ClonalFrame to our five-way genomic alignments
we obtained a phylogeny in which the two SPY
samples grouped together, as did the two SDE samples, and in which SDD and
SDE formed a clade, with SPY as an outgroup (fig.\ \ref{fig:tree5}).  This
species tree was consistent with previous results from 16S rRNA sequences
\citep{Facklam2002} and with general patterns observed genome-wide
\citep{Suzuki2011}.  The estimated levels of inter-species divergence were
substantial, at 0.48 substitution per site for SDD and SDE, and 0.66
substitutions per site for SPY and SDD/SDE.  An initial analysis with
ClonalOrigin (see \textbf{Methods}) produced an estimated per-site
population-scaled mutation rate of $\theta = $ 0.081 [interquartile range
across blocks: (0.067, 0.094)], a per-site population-scaled recombination
rate of $\rho = $ 0.012 (0.006, 0.019), and an average recombinant tract
length of $\delta = $ 744 (346, 2848) bp.  The full distributions of these
quantities across alignment blocks are shown in supplementary figure
\ref{fig:scatter3}.  A second run of the MCMC sampler produced nearly
identical results.  For comparison,
\cite{Didelot2010} obtained estimates of $\theta = 0.044$, $\rho = 0.017$,
and $\delta = 236$ for the {\em Bacillus cereus} group.  Our estimates
implied a ratio of $\rho/\theta = 0.15$, considerably lower than Didelot et
al.'s \citeyearpar{Didelot2010} estimate of $\rho/\theta = 0.405$ for {\em
  B.\ cereus}, suggesting increased rates of mutation and/or decreased
rates of recombination after controlling for differences in effective
population size.  We found that estimates of the average recombinant tract
length, $\delta$, were quite sensitive to our threshold for minimum block
length, with the inclusion of short alignments producing much larger
estimates due to edge effects.  We experimented with a range of thresholds
and selected a value (1500 bp) at which the estimates stabilized.

\subsection*{Model-based analysis of replacing HGTs}

In a second round of analysis with ClonalOrigin, we obtained samples from an
approximate posterior distribution of recombinant trees along the genome,
conditional on our five-way alignments and the previously estimated
parameters. (A recombinant tree is modeled by ClonalOrigin as the clonal
frame augmented by zero or more recombinant edges; see \textbf{Methods}.)
This distribution indicated that a majority of sites in the genome (66\%)
were topologically consistent with the clonal frame, but that there was
also substantial support for various alternative tree topologies (supplementary
table \ref{tab:Gene-tree-topologies}).  Interestingly, the second most
frequent tree topology (9.3\% of sites) had SPY and SDE as sister clades
and SDD as an outgroup, providing an initial indication of gene transfer
between SPY and SDE.  No other single tree topology appeared with a
frequency of $\geq$4\%.

To gain further insight into rates and patterns of HGT, we estimated the
rates of occurrence of various types of recombinant edges, grouping them by
their donor and recipient edges (see \textbf{Methods}).  There are 81
possible edge pairs associated with our nine-branch rooted phylogeny, but
because donor edges cannot post-date recipient edges under the model, 21 of
these pairs are prohibited, leaving 60 possible recombinant edge types.
Each of these viable edge types corresponds to a class of replacing HGTs,
including not only transfers from the donor to the recipient edge, but also
transfers from any (sufficiently old) descendant of the donor to the
recipient (due to the possibility of delayed coalescence).  Because all
recombinant edges are not equally likely \textit{a priori}, we summarized
the estimated rates by taking ratios with respect to their prior
expectations under the ClonalOrigin model.  Experiments conducted on
simulated data demonstrated that our approach had reasonable power to
detect recombination scenarios that deviated from the prior, with some
underestimation of the degree of deviation due to the use of the prior in
the inference procedure 
(see \textbf{Supplementary Material};
supplementary figures\ \ref{fig:sim2} and \ref{fig:sim3}).

We summarized these normalized rates in a heatmap (fig.\
\ref{fig:Heatmap-of-transfers}A; see also supplementary tables
\ref{tab:obsheatmap} and \ref{tab:heatmap}), and found that most cells
fell in the blue-to-white range, corresponding to rates of occurrence less
than or equal to what was expected under the prior.  A few edge types were
strongly under-represented (blue), such as those between the SDD branch and
the branches of the SPY clade ($p<0.01$ based on simulations; see
\textbf{Methods}), probably due to ecological isolation of human and
strictly veterinary pathogens.  Edges from the branch at the root were also
under-represented, suggesting a deficiency of long delays in coalescence
relative to the prior.  By contrast, the recombination edges between the
two SDE genomes, and between the two SPY genomes, were significantly
over-represented ($p<0.01$), consistent with previous evidence for abundant
intraspecies recombination in {\em Streptococcus} \citep{Feil2001}.  In
addition, we found a pronounced enrichment for recombinant edges between
the SPY and SDE clades, as has been reported previously
\citep{Kalia2001,Sachse2002,Kalia2004,Davies2005,Davies2007a}.  We also
observed a slight enrichment for SDE-to-SDD edges.

While SPY-SDE edges occurred at elevated rates in both directions, they
showed a pronounced directional asymmetry, with edges from SPY to SDE
occurring at up to four times the expected rate, while those in the reverse
direction occurred at only about twice the expected rate.  By reanalyzing
the sampled recombinant trees, we were able to obtain a Bayesian posterior
mean estimate of 2.0 (95\% credible interval: 1.8--2.2) for the ratio of the
numbers of SPY-to-SDE to SDE-to-SPY recombinant edges, compared with a ratio of 1.4
implied by the prior (the prior ratio is larger than one because of
differences in branch lengths).  In addition, we counted genes having high
probability of recombination ($>$0.6) for various types of recombinant
edges (supplementary table \ref{tab:gene-counts-replacing}), and found
that many more genes showed strong evidence of SPY-to-SDE than of SDE-to-SPY
recombinant edges (for example, 59 vs.\ 11 genes, when the entire SPY and
SDE clades are considered).  These findings provided genome-wide support
for Kalia et al.'s \citeyearpar{Kalia2001} previously unconfirmed
observations of directional asymmetry based on MLST data for seven loci.

We generated new tracks for our recently released {\em Streptococcus}
Genome Browser (http://strep-genome.bscb.cornell.edu) that summarize the
results of this model-based analysis alongside known genes, alignments, and
other annotations (fig.\ \ref{fig:ucsc}).  These tracks can be used to
inspect loci of interest and to compare the results of our model- and
parsimony-based analyses (below).  They can also be queried and intersected
with other tracks using the UCSC Table Browser.

\subsection*{Parsimony-based analysis}

To gain further insight into HGT in {\em Streptococcus}, we estimated gene
trees for 2,314 gene families from six genomes (including the SEE outgroup),
obtained parsimonious reconciliations of these gene trees with the clonal
frame (fig.\ \ref{fig:tree5}) using the Mowgli program \citep{Doyon2011},
and partitioned the inferred transfers into replacing and additive HGTs
(see \textbf{Methods}).  This analysis produced estimated numbers of five
types of events for each branch of the phylogeny (gene duplications,
losses, replacing and additive transfers, and ``appearances,'' most of
which are probably transfers from phylogenetically distant species), as
well as an estimated number of genes at each ancestral node 
(fig.\ \ref{fig:Gene-duplication-loss}).  We observed large numbers of
appearances on all branches of the tree, suggesting a steady influx of
genes into the clade by HGT.  Indeed, appearance events accounted for 1,957
of the 3,432 (57\%) gene additions (by duplication, transfer, or appearance;
loss events are excluded here).  Interestingly, the somewhat reduced
numbers of genes in SPY (particularly in SPY1) were primarily explained by
reduced rates of gene appearance, rather than by increased loss or other
factors.  The estimated numbers of ancestral genes tended to decrease
toward the root of the tree, but this likely reflects under-estimation due
to parallel losses of some genes (especially on the branches beneath the
root) rather than a true increase in gene number over evolutionary time.
An exception was the branch leading to the ancestor of the two SPY
individuals, which was enriched for gene loss events, perhaps associated
with niche adaptation.  High rates of loss also occurred on the branches to
SDE1, SDE2, and SDD.  Duplications were relatively infrequent overall, but,
as has been noted previously \citep{Marri2006}, they were substantially
enriched on external branches of the phylogeny.

% total numbers of events:
% appearance: 1957
% duplication: 257
% loss: 1293
% additive: 290
% replacement 928

We conservatively identified 290 of the 1,218 (non-appearance) transfers
(23.8\%) as additive HGTs.  As with the replacing HGTs evaluated in the
model-based analysis, these predicted additive transfers were significantly
enriched within the SDE clade, and they were weakly enriched within the SPY
clade (fig.\ \ref{fig:Heatmap-of-transfers}B and
supplementary fig.\ \ref{fig:mowgli-sim}).
They were also
significantly enriched between SDE and SDD, with a preference for the
SDE-to-SDD direction, and significantly depleted between SDD and SPY.
However, unlike the replacing HGTs from the previous section, these
additive transfers were not significantly enriched between SPY and SDE.
They also did not exhibit a pronounced directional asymmetry between SPY
and SDE, except between the ancestral SPY and SDE branches, where they were
significantly depleted in both directions, but more strongly from
SDE-to-SPY.  It is worth noting, however, that this apparent asymmetry in
depletion could in fact reflect an asymmetry in enrichment, as the
normalization in this case reflects average rates across the tree, and an
inflated average could cause a global shift in the heatmap toward depletion
(blue).  Other differences also make it difficult to compare the model- and
parsimony-based analyses---for example, the first considered the core
genome only, while the second considered the pan genome; and the two types
of analysis might have quite different power for events affecting different
portions of the phylogeny.  Nevertheless, we find much clearer evidence
that homologous recombination, rather than additive transfer, has driven
the apparent gene flow between SPY and SDE, particularly in the SPY-to-SDE
direction.  Notably, we compared the replacing HGTs identified in the
parsimony-based analysis with those from the model-based analysis, and
found reasonable concordance, despite substantial differences between the
data sets and methods
(see \textbf{Supplementary Material}).

\subsection*{Functional categories of transferred genes}

To gain insight into the functional impact of HGT, we assigned genes to
various functional categories and looked for statistical associations
between these categories and predicted gene transfer events (see
\textbf{Methods}).  First, we partitioned all genes into ``virulence'' and
``non-virulence'' categories, considering genes whose homologs in other
genera of bacteria exhibit virulence phenotypes (based on VFDB) to be
virulence genes \citep[see][]{Suzuki2011}.  Among gene families for which
we inferred additive HGTs, virulence genes were significantly enriched
($p=1.33 \times 10^{-11}$), possibly due to selection favoring
virulence-related functions.  Interestingly, virulence genes associated
with additive HGTs showed a pronounced enrichment for SPY-to-SDE events
(25.0\% of events compared with 12.1\% for all genes at the level of the
entire SPY and SDE clades, $p=0.01$, Fisher's exact test; supplementary
table \ref{tab:gene-counts-additive}). 
% this is based on a one-sided test of the follow 2x2 table
%     [,1] [,2]
%[1,]   10   30
%[2,]   22  203
Transfers in the reverse direction (SDE-to-SPY) were also over-represented
(17.5\% vs.\ 10.6\%), but not significantly ($p=0.11$).  Replacing gene
transfers, on the other hand, were not significantly enriched for virulence
genes, and those virulence genes that did exhibit strong evidence of
replacing transfers were not enriched for SDE/SPY transfers (supplementary
table \ref{tab:gene-counts-replacing}).  We also found a significant
association between virulence genes and gene duplications ($p=7.13 \times
10^{-3}$).  

Next, we assigned gene ontology (GO) terms to all genes, and searched for
significant associations with transfer events.  Several GO categories
having to do with transposition, recombination, and DNA integration were
significantly associated with additive gene transfers (table
\ref{tab:go-events}), consistent with observations in other bacterial
groups \citep[e.g.,][]{Liu2009}.  Replacing transfers, on the other hand,
showed much weaker, and quite different, functional associations compared with
additive transfers.  These associations were similar for replacing
transfers inferred by the model- and parsimony-based methods (supplementary
tables \ref{tab:functional} and \ref{tab:go-events-recombining}).


\subsection*{Positive selection and gene transfer}

% some of this should really go in "Methods".  Should also explain exactly
% how the MWU tests are done

Finally, we tested for evidence of positive selection within the gene
families, and for associations between positive selection and various types
of events.  We performed a likelihood ratio test using PAML's ``sites
models'' of M1a and M2b \citep{Yang2007}
on each of the 2,991 gene-fragment families with three or more genes.
After controlling for multiple comparisons, 31
families showed evidence of positive selection (FDR$<$5\%).  Among
high-confidence gene trees (bootstrap values of $>$90\% on every branch),
families with duplication events were slightly enriched for evidence of
positive selection ($p<0.046$, Mann-Whitney $U$ test [MWU]) as were families
with additive transfers ($p<0.002$, MWU).  Of the 31 families with
significant positive selection, 13 families were part of the core genome
analyzed by ClonalOrigin.  Replacing transfers from SDE to SPY and from SPY
to SDE were slightly enriched for positive selection ($p<0.005$ and
$p<0.01$, MWU).  For 11 of the 13 (85\%) families showing evidence of
positive selection, the SPY-to-SDE direction had greater intensity than the
SDE-to-SPY direction, consistent with overall directional bias (83.3\%).
The same directional bias was not seen for additive transfers between the
two clades.


% we listed several virulence genes with high ranks (Table S\ref{tab:virhgt}).
% The whole gene of  putative dipeptidase (SpyM3\_0465) was transferred, which was
% a exemplary case for the finding \citep{Chan2009} that virulence genes were transferred as a whole
% unit.  Another example of whole gene transfer was superoxide
% dismutase (SpyM3\_1071).  We investigated the unit of recombining gene transfer
% by finding 58 genes for which a recombinant edge of certain types were sampled
% with probability at least 0.9 by focusing on specific types of recombinant edges
% (Table S\ref{tab:genes-transfer}).  In some cases, parts of a gene were inferred
% to have been transferred, for instance, transfer from SPY1 to SDE1 was inferred
% for 29\% of the aspartyl-tRNA synthetase gene. Recombinant tracts often were
% shown to span two genes.  In other cases, the inferred transfer units spaned a
% set of as much as eight adjacent genes: loci from SpyM3\_1518 to SpyM3\_1525.

\section*{Discussion}

There has been a great deal of interest for decades in the process of
horizontal gene transfer (HGT) and the manner in which it has influenced
phylogenetic relationships, particularly among bacteria \citep{Koonin2001}.
Most previous studies, however, have either simply examined patterns of
phylogenetic discordance, without differentiating between additive and
replacing HGTs \citep[e.g.,][]{Lerat2005}, or have focused specifically on
the process of homologous recombination, corresponding to replacing HGTs
\citep{Didelot2007,Didelot2010}.  To our knowledge, this is the first study
to examine both processes on a genome-wide scale, using both model- and
parsimony-based methods.

We have focused our analysis on {\em Streptococcus pyogenes} (SPY) and two
of its close relatives, {\em S.\ dysgalactiae} ssp.\ {\em equisimilis}
(SDE) and {\em S.\ dysgalactiae} ssp.\ {\em dysgalactiae} (SDD), a human
commensal-like organism and a strict veterinary pathogen, respectively.
Our combined analysis reveals strong evidence of gene flow both within and
between the SPY and SDE groups.  Interestingly, events between SPY and SDE
are more prominent among inferred replacing than among inferred additive
transfers, suggesting that they may be driven by homologous recombination,
although we acknowledge several challenges in comparing inferences from the
model- and parsimony-based analyses.  We also find strong support for an
asymmetry in replacing gene transfers between SPY and SDE, with a
preference for the SPY-to-SDE direction.  The situation with SDD is more
complex, with somewhat elevated rates of SDE-to-SDD transfer, more
pronounced for additive than for replacing HGTs, mixed evidence for
SDD-to-SDE transfer, and generally reduced rates of transfer between SDD
and SPY, perhaps owing to a combination of genetic divergence and different
ecological niches.  We also find that virulence genes are significantly
associated with additive, but not replacing, gene transfers, as are genes
under positive selection.  Genes that have undergone replacing transfers
between SPY and SDE are also enriched for positive selection.  A central
component of our study is the use of simulations to demonstrate that our
methods have good power and accuracy in the detection of both types of
transfer events.

The finding of abundant, asymmetric gene flow between
SPY and SDE raises a number of questions.
First, can Kalia et al.'s \citeyearpar{Kalia2001} observations be
confirmed using our data, despite the authors' inability to reproduce them?
%\texttt{CHECK:} 
Six of their seven loci were represented in our five-way alignments
(\textit{gki}, \textit{gtr}, 
\textit{murI}, \textit{mutS}, \textit{recP}, and \textit{xpt}) 
and we did find 
strong evidence of recombination two of them 
(\textit{recP} and \textit{xpt}) and weak evidence in two others
(\textit{gki} and \textit{mutS}).  
These results were generally consistent those of \citet{Kalia2001}, except
that they found no recombination in {\em mutS}.
%Our visual inspection of the recombination probability on 
%\textit{recP} and \textit{xpt} indicated that 
%the replacing gene transfer appeared to be biased towards
%the directions from SPY to SDE.
%that the six of their seven loci represented in our five-way alignments
%(\textit{gki}, \textit{gtr}, 
%\textit{murI}, \textit{mutS}, \textit{recP}, and \textit{xpt}) showed
%consistent results with those of \citet{Kalia2001}
% In S. pyogenes MGAS315 genome track, we can check
% gki -> not in the region but its neighboring
% chr1:1207587-1208084
% /locus_tag="SpyM3_1180"
% 
% gtr -> no recombination 
% chr1: 1186522-1186073
% /locus_tag="SpyM3_1160"
%
% murI -> no recombination
% chr1:303215-303652
% /locus_tag="SpyM3_0262"
%
% mutS -> a little bit
% chr1:1836841-1837245
% /locus_tag="SpyM3_1806"
%
% recP -> somewhat strong (consistent with Kalia2001's conclusion)
% chr1:1454877-1455335
% /locus_tag="SpyM3_1462"
%
% xpt -> somewhat strong
% chr1:846784-847233
% /locus_tag="SpyM3_0794"
%
% yqil -> no data available
% chr1:131266-131699
% /locus_tag="SpyM3_0108"
%
%Of course, our study differs from theirs in numerous respects, but this
%suggests the two loci are prone to recombination.
%
Second, are the observed patterns of gene flow truly driven by replacing
transfers, or could additive transfers also be contributing?  Here, our
evidence is weaker, because our parsimony-based method has limited ability
to resolve the direction of transfer.  
% FIXME: THIS NEEDS WORK
More genomes will be needed to
evaluate the directionality of additive transfers between these groups.
Third, what could explain the observed asymmetry in gene flow?  Possible
mechanisms include differences in the sizes or demographic structure of
co-mingling populations of cells, or differences in barriers to gene
transfer \citep{Thomas2005}.  \cite{Kalia2001} speculated that 
SDE might be inherently less capable than SPY to acquire DNA by homologous
recombination, for example, due to differences in the 
restriction-modification system (RMS), hetero-duplex formation, and/or mismatch
repair.  
Interestingly,
while most of \textit{S.\ pyogenes} genome contain both type I and type II
RMSs, the two SDE genomes studied here contain
only type II RMSs \citep{Roberts2010}.
%  (THIS STILL NEEDS SOME WORK)

% can we integrate some of this?  reread Kalia
%More recently, 
%As biological roles of RMS might be to maintain and
%control species identity \citep{Jeltsch2003}, \cite{Budroni2011a} recently
%studied that \textit{Neisseria meningitidis} 
%phylogenetic clades were associated with RMS, claiming that RMS modulated
%homologous recombination in the bacteria of \textit{N.\ meningitidis}.

%Most of \textit{S.\ pyogenes} genomes contained restriction-modification
%type I system as well as type II system. The two genomes of SDE contained
%only type II system. A more robust self-defense system in SPY could have
%made the species less capable of acquiring foreign genetic material from
%SDE.
% can we follow-up on this in some way?

%More SDE genomes might be beneficial to future studies. 


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

In comparing replacing and additive HGTs, it is worth bearing in mind that
these events are likely to be strongly correlated with the core and
dispensable portions of the genome.  In part, this reflects an
ascertainment bias in the inference procedures used to detect these events.
Methods for detecting recombination (replacing transfers) tend to rely on
large-scale alignments spanning multiple syntenic loci in order to gain
power.  As a result, these methods are effectively limited to application
in the core genome, even though homologous recombination probably also
occurs between genes that are less widely shared across species.  By
contrast, phylogenetic reconciliation methods can be applied to the entire
pan genome, as in this study.  However, this correlation is also likely to
hold for true events.  True additive transfer events will be enriched in
the dispensable genome, by definition, and it is likely that homologous
recombination occurs at higher rates in the core genome (which is enriched
for close homologs between species).  For all of these reasons, some
apparent differences between the two types of transfer events are likely to
reflect general differences between the core and dispensable portions of
the genome.

As sequence data becomes available for many more closely and distantly
related organisms, it will become increasingly important to devise improved
methods that consider both additive and replacing HGT.  As noted above, the
current statistical models of recombination, such as ClonalOrigin, are
effectively limited to considering orthologs, with one copy per species, in
the core genome.  The phylogenetic reconciliation methods, on the other
hand, fail to allow for multi-locus gene transfer events, and ignore
information from branch lengths.  We have shown that it is possible, to a
degree, to extend these methods to distinguish between additive and
replacing transfers, but our methods are limited by their dependency on
heuristic rules and parsimony assumptions.  It may be possible to develop
integrated statistical models that consider both types of gene transfers,
and produce unbiased estimates of the relative rates at which these events
occur.  Another extension worth considering is direct modeling of
incomplete lineage sorting (ILS), which is likely to be increasingly
important as larger phylogenies are considered, especially given the large
effective population sizes of many bacteria.  ILS has recently been
integrated into models for duplication and loss \citep{Rasmussen2012} but
has yet to be considered together with gene transfer.  
%It is permitted by
%ClonalOrigin, but only in an 
%indirect and approximate manner, through the introduction of recombination
%edges to ancestral branches (which can be misinterpreted as evidence of
%recombination).  
We expect that the combination of improved models and richer data sets
will allow for a much more detailed understanding of the important
process of horizontal gene transfer.



%Molecular evolution of bacterial genomes is multifaceted. Researches of
%bacterial evolution often focused on one side of many evolutionary processes.
%While understanding one aspect of bacterial evolutionary processes we could be
%ignorant of the other features.  Horizontal gene transfer process was often
%considered in macro-evolutionary scale using a great deal of bacterial species.
%As closely related species or groups of individual bacteria are becoming a focus
%of evolutionary study, recombining gene transfer process would surface as
%another facet of evolutionary forces.  Because the two gene transfer processes
%would act on closely related bacterial genomes simultaneously, ignoring one of
%the two processes in studying evolution of bacteria would lead to a narrowed
%point of view to the underlying evolutionary forces. Our study of SPY and SDE
%genomes undertook the work of discriminating recombining and horizontal gene
%transfer processes because the two species were very closely related.  In this
%study, we showed that the two gene transfer processes acted on different
%biological functions.  Recombining gene transfer process in the analysis of SPY
%and SDE genomes was related with genes encoding membrane-bound molecules and
%aminoacyl tRNA synthetases. On the contrary, horizontal gene transfer process
%was found to be associated with transposases, which would allow horizontal gene
%transfer of their neighboring genes.  Because SPY was a human-specific pathogen,
%and SDE was an opportunistic human pathogen, we were interested in association
%of pathogenicity with the two gene transfer processes. We also showed that
%virulence genes were more associated with horizontal gene transfer process than
%with recombining gene transfer process.  With the current data of only SPY and
%SDE genomes we could not conclude whether the association of virulence genes
%with horizontal gene transfer not recombining gene transfer was applicable
%generally to cases of other species.  However, without the discrimination of the
%two processes it would have been difficult to find the disparate association of
%biological functions or virulence genes with the different gene transfer
%processes. 
%
%We showed differences in the profile of functional category association of the
%two gene transfer processes thanks to the approach of dicriminating the two gene
%transfer processes.  Sugar phosphotransferases could be important in signal
%transduction to recognize environmental changes.  Transporters, and fatty acid
%and lipid biosyntheses would be involved in membrane-bound molecules.  Because
%bacteria as a unicellular organism should be able to promptly respond to changes
%to environment, some evolutionary pressures must have been acting on genes
%encoding membrane-bound molecules.  We also found that the recombination
%intensity and requency of recombining gene transfer events would be useful in
%associating several aminoacyl tRNA synthetases genes belonging to gene ontology
%term of translation (GO:0006412) with recombining gene transfer. It had been a
%focus of gene transfer in bacterial species \citep{Woese2000}.  The biological
%role of transposases was concordant with horizontal gene transfer process
%because the process involved in transposases would allow bacteria to acquire
%foreign genes without the requirement of homologous regions in the bacteria.  
%
%Another contribution of our study to bacterial gene transfer research is to
%elucidate the net directionality of gene flow from SPY to SDE in the core
%genomes. Although the issue of net directionality has been tackled, our study
%was the first attempt of resolving the issue in genome-wide scale. Yet, it still
%remains to be seen whether this net directionality of gene flow from SPY to SDE
%happened in the pan genome. A model-based method with more refined models would
%be necessary because the parsimony-based method we employed simply could not
%show any net directionality between SPY and SDE in the pan genome. Although we
%are unsure whether much larger number of genomes could be useful in finding net
%directionality, more SDE genomes might be beneficial to future studies. 
%
%Were there unbalanced recombining gene transfer between the two species, what
%were the barriers \citep{Thomas2005} to cause the unbalanced gene transfer in
%the species?  Kalia et al.\ \cite{Kalia2001} discussed restriction-modification
%system (RMS) as a possible mechanism of the unbalanced gene flow.
%Restriction-modification system is a biological self-defense mechanism of
%bacteria, which is diverse in prokaryote world.  As biological roles of RMS
%might be to maintain and control species identity \citep{Jeltsch2003}, Budroni
%et al.\ \cite{Budroni2011a} recently studied that \textit{Neisseria
%meningitidis} phylogenetic clades were associated with RMS, claiming that RMS
%modulated homologous recombination in the bacteria of \textit{N.\ meningitidis}.
%Most of \textit{S.\ pyogenes} genomes contained restriction-modification type I
%system as well as type II system. The two genomes of SDE contained only type II
%system. More tight self-defense system of SPY may have made the species less
%capable of aquiring foreign genetic material from SDE \citep{Kalia2001}.
%
%We reexamined the seven genes that Kalia et al.\ \cite{Kalia2001} had studied
%for net directionality of gene transfer between SPY and SDE.  The internal
%fragments of the six housekeeping genes (\textit{gki}, \textit{gtr},
%\textit{murI}, \textit{mutS}, \textit{recP}, \textit{xpt}) were located in the
%core genome that we investigated in this study.  Except for \textit{mutS} where
%\cite{Kalia2001} did not find recombination, our results were
%consistent with those of \cite{Kalia2001}.  
%\cite{Pinho2010} reported recombination in the gene \textit{parC}, in which we
%also observed moderate recombination.
%
%It remains to be seen whether the inability of detecting net directionality
%between SPY and SDE in the pan genomes was because of lack of power of the
%parsimony-based approach that we employed. Considering gene duplication, loss,
%and horizontal gene transfer in a model-based approach would be desirable in the
%coming ages of abundant bacterial genomes. The model might consider genome
%alignments where each taxon does not necessarily have a unique gene; some taxa
%might have no gene due to gene loss, and other taxa multiple genes due to gene
%gain, duplication, or horizontal gene transfer. With these types of models will
%we be able to address more interesting evolutionary questions in bacteria. This
%need of more refined models would be also true even for the study of net
%directionality between species in the core genome.  In other words, one could
%envision a model where a population two-taxa (i.e., SPY and SD groups) tree is
%superimposed on the species tree shown in Figure S\ref{fig:clonalorigin}A. A
%splitting event of the population tree could happen somewhere between the root
%of the species tree and the internal node of SPY. Recombinant edges that connect
%branches within SPY or SD clades would be more frequent than those that connect
%branches between SPY and SDE clades.  With the reference to the recombination
%within a clade one parameter can dictate the degree of gene from from SPY to
%SDE, and another for the reverse direction.  If the two parameters could be
%incorporated in the model, then they might be used to compare the degrees of
%gene flow between the two species in a more statistically sound way.  We hope
%that we will see this refinement of the model to better illustrate bacterial
%evolution in forthcoming bacterial genomic era.

% ACS: Let's let the journal do this 
%\section*{Supplementary Material}
%Supplementary figures S1-S10 and tables S1-S9 are available at Molecular Biology
%and Evolution online (http://mbe.oxfordjournals.org/).

% Do NOT remove this, even if you are not including acknowledgments
\section*{Acknowledgments}

This study was funded by the National Institute of Allergy and Infectious
Disease, US National Institutes of Health, under grant number AI073368-01A2
(to M.J.S.\ and A.S).  Additional support was provided by National Science
Foundation CAREER Award DBI-0644111 and a David and Lucile Packard
Fellowship for Science and Engineering (to A.S.).  We thank Xavier Didelot
for assistance with ClonalOrigin and Haruo Suzuki for providing data for
SDD.

%\section*{Author Contributions}

%Conceived and designed the experiments: SCC MDR IG MJS AS.
%Performed the experiments: SCC MDR MJH.
%Analyzed the data: SCC MDR MJH IG.
%Contributed reagents/materials/analysis tools: MJS.
%Wrote the paper: SCC IG AS (with review and contributions from all authors).
%\clearpage

% The bibtex filename
\renewcommand*{\refname}{Literature Cited}
\bibliographystyle{../../latex/bst/mbe}
\bibliography{siepel-strep}
\clearpage{}


\section*{Tables}
%\begin{table}[!ht]
%\caption{
%\bf{Table title}}
%\begin{tabular}{|c|c|c|}
%table information
%\end{tabular}
%\begin{flushleft}Table caption
%\end{flushleft}
%\label{tab:label}
% \end{table}

%=============================================================================
\begin{table}[!h]

\caption{{\bf Gene Ontology (GO) enrichments for additive gene
  transfers}}
\vspace{1ex}

\noindent \begin{centering}
\begin{tabular}{ccccl}
\hline 
$p^a$ & $q^b$ & Count$^c$ & GO term & Description \\
\hline 
7e-61 & 3e-58 & 67 & GO:0006313 & transposition, DNA-mediated\\
3e-58 & 7e-56 & 59 & GO:0004803 & transposase activity\\
1e-30 & 2e-28 & 52 & GO:0015074 & DNA integration\\
1e-28 & 1e-26 & 23 & GO:0032196 & transposition\\
1e-13 & 1e-11 & 527 & GO:0003677 & DNA binding\\
2e-12 & 2e-10 & 86 & GO:0006310 & DNA recombination\\
1e-11 & 7e-10 & 197 & GO:0003676 & nucleic acid binding\\
3e-09 & 1e-07 & 36 & GO:0016987 & sigma factor activity\\
3e-09 & 1e-07 & 36 & GO:0006352 & transcription initiation\\
3e-08 & 1e-06 & 142 & GO:0043565 & sequence-specific DNA binding\\
4e-07 & 1e-05 & 12 & GO:0008170 & N-methyltransferase activity\\
1e-06 & 4e-05 & 12 & GO:0019867 & outer membrane\\
2e-06 & 5e-05 & 25 & GO:0007059 & chromosome segregation\\
1e-05 & 4e-04 & 15 & GO:0006306 & DNA methylation\\
2e-05 & 6e-04 & 15 & GO:0006470 & protein amino acid dephosphorylation\\
2e-04 & 4e-03 & 76 & GO:0005618 & cell wall\\
3e-04 & 7e-03 & 25 & GO:0008236 & serine-type peptidase activity\\
3e-04 & 7e-03 & 65 & GO:0009986 & cell surface\\
5e-04 & 1e-02 & 10 & GO:0003872 & 6-phosphofructokinase activity\\
1e-03 & 2e-02 & 11 & GO:0009307 & DNA restriction-modification system\\
2e-03 & 5e-02 & 84 & GO:0006974 & response to DNA damage stimulus\\
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$$P$-value based on a Mann-Whitney $U$ test (Methods).\\
$^b$Corresponding false discovery rate estimated by the
Benjamini-Hochberg method.  All categories having at least ten genes and
$q\leq 0.05$ are displayed.\\ 
$^c$Number of genes assigned to category.\\
\end{flushleft}
\label{tab:go-events}
\end{table}
\clearpage{}

%=============================================================================

\section*{Figures}

%=============================================================================
\begin{figure}[!h]
\begin{center}
\includegraphics[width=3in]{figures/cells.pdf}
\end{center}
\caption{ {\bf Replacing and additive horizontal gene transfer.}  (A)
  Foreign DNA can be integrated into a recipient genome by homologous
  recombination (a {\em replacing transfer}, left) or additive integration
  (an {\em additive transfer}, right). (B) These types of transfers produce
  distinct phylogenetic signatures.  In homologous recombination a segment
  of DNA is effectively overwritten by a homologous segment from another
  species, which causes a lineage in the phylogeny to be replaced by a
  transferred lineage.  This type of transfer can be identified in
  gene-tree/species-tree reconciliation by the appearance of coinciding
  transfer and loss events (left tree).  Additive integration, on the other
  hand, leads to a transfer event that is not paired with a loss event
  (right tree).  Of course, parallel losses can cause an additive transfer
  to appear similar to a replacing transfer.  In our analysis, we
  conservatively require that at least one descendant species (here, 
    $B$) contains genes that both descend from ($b_1$) and do not descend
  from ($b_2$) a lineage predicted to have been transferred when
  inferring an additive transfer event.}
\label{fig:hgt}
\end{figure}
\clearpage{}%
%=============================================================================

%=============================================================================
\begin{figure}%[!ht]
\noindent \begin{centering}
\includegraphics[scale=0.5]{figures/cornellf-3-tree}
\par\end{centering}
\caption{ {\bf Clonal frame inferred for the five genomes.}  This phylogeny
  was inferred using the ClonalFrame program \citep{Didelot2007}.  Branch
  lengths are in units of expected substitutions per site and are drawn to
  scale in the horizontal dimension.  The labels for the ancestral nodes of
  the tree and the branches immediately ancestral to them (SPY, SDE, and
  SD) are used throughout the paper.  The outgroup species, SEE, was not
  part of the estimated clonal frame but was used in the parsimony-based
  analysis.}
\label{fig:tree5}
\end{figure}
%=============================================================================

% A transfer is a additive if there exists an extant descendant species
% (e.g. C and D) that contains both descendants (bold) and
% non-descendants (normal weight) of the transferred gene (black
% circle).


%=============================================================================
\begin{figure}
% \includegraphics[scale=0.45]{figures/heatmap-recedge}
\begin{center}
\includegraphics[scale=0.45]{figures/heatmap}
\caption{\label{fig:Heatmap-of-transfers}
{\bf Heatmaps showing rates of replacing and additive transfers.}
Each cell of the heat map represents the base-2 logarithm 
of the ratio of the estimated number recombination events to its prior
expectation, for the corresponding donor ($y$ axis) and recipient ($x$
axis) branches.  
Cells in black indicate prohibited transfer events.
%
(A) Replacing transfers, as inferred by the model-based approach. The
plotted values represent 
average values across sampled recombinant
trees. The prior considers the clonal frame with branch lengths (fig.\
\ref{fig:tree5}) and the 
global estimates of the three population parameters. Prohibited transfer events
are ones for which the recipient branch is strictly older than the donor
branch. An asterisks indicates statistical significance ($p<0.01$).
%
(B) Additive transfers, as inferred by the parsimony-based
approach. The plotted log ratios reflect
total numbers of inferred additive transfers across all gene families.
The prior considers the clonal frame (with branch lengths) and the total number
of events of each type inferred by the analysis. Prohibited transfer events are
ones for which the recipient and donor branch do not share a common time
interval.
branch. Significance levels are denoted by asterisks: one for $p<0.01$, two for
$p<0.005$, and three for $p<0.001$.  See {\bf Methods} for details.}
\end{center}
\end{figure}
\clearpage{}
%=============================================================================

%=============================================================================
\begin{figure}%[!ht]
\begin{center}
\includegraphics[scale=0.5]{figures/ucsc}
% used genomic interval chr1:511,000-513,500
\caption{ {\bf Genome Browser tracks.}  Our inferences of replacing and
  additive horizontal gene transfers are summarized in new tracks in the
  {\em Streptococcus} Genome Browser \citep{Suzuki2011}.  (A) The main
  browser display, with gene annotations for the selected reference genome
  ({\em S.\ pyogenes} strain MGAS315) shown in red and putative virulence
  genes highlighted in blue. The third track from top (in black) shows the
  putative non-recombining gene fragments analyzed by Mowgli.  The next series
  of tracks shows the results of the ClonalOrigin analysis of
  replacing gene transfers.  Each track in this series corresponds to a
  recipient lineage in the phylogeny (fig.\ \ref{fig:tree5}) and describes
  the posterior probabilities along the genome of recombinant edges from
  all possible donor lineages (shown in different colors; see key at
  right).  Here the putative virulence gene SpyM3\_0465, a dipeptidase,
  shows strong evidence of a recombinant edge from the SPY to the 
  SDE lineage, as well as some evidence of a SPY$\rightarrow$SPY2 edge.
  The genome-wide multiple alignment obtained with Mauve is shown at
  bottom.  (B) The gene tree that is displayed after clicking on the
  highlighted gene fragment.  The tree estimated by RAxML is consistent
  with the combined influence of SPY$\rightarrow$SDE and
  SPY$\rightarrow$SPY2 replacing transfers, as inferred by ClonalOrigin.}
\label{fig:ucsc}
\end{center}
\end{figure}
\clearpage{}%
%=============================================================================

%=============================================================================
\begin{figure}
\begin{center}
\includegraphics[width=5in]{figures/strep-events}
\caption{\label{fig:Gene-duplication-loss} {\bf Distribution of inferred
    gene duplication, loss, and transfer events across the six-species
    phylogeny.}  Numbers at nodes represent gene counts for extant species,
  and estimated counts for ancestral species.  Numbers on branches indicate
  inferred gene appearances (A*), duplications (D*), losses (L*), additive
  transfers (T*), and replacing transfers (R*).  Transfer events are
  recorded on recipient branches (donors are not indicated).}
\end{center}
\end{figure}
\clearpage{}%
%=============================================================================


\end{document}

