% Four potential reviewers whose e-mail addresses you can provide
% Debra Bessen
% Claus O. Wilke
% Didelot Xavier
% Bruce R. Levin
% Michael Wessels

% Template for PLoS
% Version 1.0 January 2009
%
% To compile to pdf, run:
% latex plos.template
% bibtex plos.template
% latex plos.template
% latex plos.template
% dvipdf plos.template

\documentclass[10pt]{article}

% amsmath package, useful for mathematical formulas
\usepackage{amsmath}
% amssymb package, useful for mathematical symbols
\usepackage{amssymb}

% graphicx package, useful for including eps and pdf graphics
% include graphics with the command \includegraphics
\usepackage{graphicx}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

\usepackage{color} 

% Use doublespacing - comment out for single spacing
\usepackage{setspace} 
\doublespacing

% FIXME: Remove this package when submitting the manuscript.
\usepackage{url}

% Text layout
\topmargin 0.0cm
\oddsidemargin 0.5cm
\evensidemargin 0.5cm
\textwidth 16cm 
\textheight 21cm

% Bold the 'Figure #' in the caption and separate it with a period
% Captions will be left justified
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

% Use the PLoS provided bibtex style
% \bibliographystyle{plos}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}

% FIXME: Remove the followings if you have to when submitting the manuscript.
\let\citep\cite
\let\citet\cite
\providecommand{\tabularnewline}{\\}
\newcommand{\lyxdot}{.}
\@ifundefined{showcaptionsetup}{}{%
 \PassOptionsToPackage{caption=false}{subfig}}
\usepackage{subfig}

\makeatother


% Leave date blank
\date{}

\pagestyle{myheadings}
%% ** EDIT HERE **
\markright{Gene Transfer using Streptococcus Genomes}

%% ** EDIT HERE **
%% PLEASE INCLUDE ALL MACROS BELOW
%% END MACROS SECTION

\begin{document}

% Title must be 150 characters or less
\begin{flushleft}
{\Large
\textbf{Inferring homologous recombination and horizontal gene transfer 
using whole genomes of \textit{Streptococcus pyogenes} 
and\textit{ S.\ dysgalactiae} ssp.\textit{\ equisimilis}}
}
% Short Title (running head): Gene transfer using Streptococcus genomes
% Insert Author names, affiliations and corresponding author email.
\\
Sang Chul Choi$^{1}$, Matthew D. Rasmussen$^{1}$, 
Melissa Jane Hubisz$^{1}$, Ilan Gronau$^{1}$,
Michael J. Stanhope$^{2}$, and Adam Siepel$^{1,\ast}$
\\
\bf{1} 
Department of Biological Statistics and Computational Biology,
Cornell University, Ithaca, NY 14853, USA
\\
\bf{2}
Department of Population Medicine and Diagnostic Sciences,
College of Veterinary Medicine, Cornell University, Ithaca, NY 14853, USA
\\
$\ast$ E-mail: acs4@cornell.edu, Phone: 607-254-1157, Fax: 607-255-4698
\end{flushleft}

% \section*{Things To Do for submitting this to PLoS Pathogens}
% \begin{enumerate}
% \item Legends for the main figures as a separate section
% \item We can add at least 50 words in section \textbf{Abstract}.
% \item Figures, tables, and supporting information need to be ready.
% \item Merge Genome Browser tracks to the official one.
% \item Publish source codes at public domain.
% \end{enumerate}
%
% \section*{A list of our major and minor findings or observations}
% \begin{enumerate}
% \item We showed that recombining gene transfer from SPY to SDE was strong
% enough.
% \item A few evidences for the net directionality from SPY to SDE:
% time-constrained gene transfer between SPY to SDE also showed the directionality
% from SPY to SDE.
% \item We showed that virulence genes were associated with horizontal gene
% transfer. 
% \item We found that recombining gene tranfer processes are involved in
% membrane-bound protein encoding genes and aminoacyl tRNA synthetases.
% \item We found that horizontal gene transfer processes are involved in
% transposases.
% \item We provided evidence that recombination process affected more sites along
% the core genome of SPY and SDE than mutation process.
% \item We found a recombination hotspot at 200 kb from the origin of replication
% along the genome of SDE1.
% \item We showed that species-tree-topology-changing recombinant edges were more
% reliably recovered than species-tree-topology-not-changing edges.
% \item We showed that recombination intensity might be useful in finding genes
% under high recombining gene transfer.
% \item We found that SPY and SDE are more closely related than SDD and SDE in
% about 10\% of the core genome.
% \item Recombination intensity and number of recombining gene transfer events
% were positively correlated.
% \item We found reduction of numbers of genes after divergence of SDE and SPY.
% \item We confirmed the species tree that relates the five genomes.
% \item We listed several virulence genes associated with recombining and
% horizontal gene transfer.
% \end{enumerate}
% 
% \section*{Suggestions about the manuscript}
% \begin{enumerate}
% \item Ilan thought that a species tree must be shown at the outset of the
% section \textbf{Results}.
% \item Can selections after gene transfer be tested?
% \end{enumerate}

% Please keep the abstract between 250 and 300 words
\section*{Abstract}

Homologous recombination and horizontal gene transfer are investigated with
complete genomes of closely related bacterial species of \textit{Streptococcus
pyogenes} (SPY) and \textit{S.\ dysgalactiae} subspecies \textit{equisimilis}
(SDE).  The former is a pervasive human-specific pathogen, and the latter an
opportunistic human pathogen.  We aims at which parts of the genomic
organization and human pathogenic virulence factors of SDE could have originated
in those of SPY, or vice versa.  Two mechanisms of gene transfer in bacteria
were known; recipient bacteria of transferred genes can substitute transferred
genes for homologous parts of their genomes (Recombining Gene Transfer), or
integrate transferred genes into their genomes without replacement (Horizontal
Gene Transfer).  We discriminated the two processes by using a model-based
approach for studying recombining gene transfer, and using a parsimony-based
approach for both transfer processes.  Our genome-wide survey of gene transfer
indicated that gene transfer via homologous recombination occurred more often
for the net direction from SPY to SDE.  We also found evidence that virulence
genes transferred more often via non-recombining fashion without replacing parts
of destination genomes than did non-virulence genes.  
A genome-wide resource of
recombining and non-recombining gene transfer for the study of the SPY and SDE
genomes is provided so that research community in the pathogenic SPY and SDE
bacteria could take advantage of it. 


% Please keep the Author Summary between 150 and 200 words
% Use first person. PLoS ONE authors please skip this step. 
% Author Summary not valid for PLoS ONE submissions.   
\section*{Author Summary}

\textit{Streptococcus pyogenes} (SPY) has afflicted people around the globe
causing many diseases. Its closely related species, \textit{S.\ dysgalactiae}
subspecies \textit{equisimilis} (SDE), used to be known as veterinary pathogens
have been reported to share disease profiles with SPY.  While gene transfers
between the two species were shown in small portions of genomic loci, no study
on the gene transfer at genome-scale has been performed because of lack of
methodologies implemented or genomic data available. Recent advances in both
method development and genomic data allowed to study gene transfer between the
two species in genome scale.  Recipient bacteria of transferred genes can
substitute transferred genes for homologous parts of their genomes, or integrate
transferred genes into their genomes without replacement.  The former is called
recombining gene transfer, and the latter horizontal gene transfer. By
discriminating the two processes, We found that virulence genes were more
associated with horizontal gene transfer than recombining gene transfer.  We
also found a signficant signal of recombining gene transfer for the net
direction from SPY to SDE in the core genome.  Our study provides new insights
into gene transfer between SPY snd SDE in genome scale.  

\section*{Introduction}

Bacteria infect humans with a variety of diseases: e.g., food poisoning
(\textit{Eschericha Coli}), strep throat (\textit{Streptococcus pyogenes}),
tuberculosis (\textit{Mycobacterium tuberculosis}), cholera (\textit{Vibrio
cholera}), anthrax (\textit{Bacillus anthracis}), pneumonia
(\textit{Chlamydophila pneumoniae}), and influenza (\textit{Haemophilus
influenzae}), to name a few.  \textit{Streptococcus pyogenes} (SPY) belonging to
group A streptococci (GAS) primarily infect humans in the throat and skin
\citep{Broyles2009}. It has high prevalence throughout the world causing
diseases ranging from mild illnesses such as pharyngitis to severe invasive
streptococcal diseases including necrotizing fasciitis and toxic shock syndrome
\citep{Cunningham2000a}.  Its close relative, \textit{Streptococcus
dysgalactiae} subspecies \textit{equisimilis} (SDE) belonging to group C or G
streptococci (GCS/GGS), used to be generally regarded as a veterinary pathogen
\citep{Vandamme1996}. However, it has been reported to cause serious and
life-threatening streptococcal diseases to humans traditionally associated with
GAS \citep{Brandt2009}.  Because of the overlap in disease profile and
ecological niche of tissue sites betwen SPY and SDE, these bacteria may as
well interact with each other by exchanging parts of their genomes.  

Therefore, the research topic on gene exchange between SPY and SDE has been
performed in various ways. Sachse et al.\ \citet{Sachse2002} showed gene transfer from SDE to
SPY for the downstream of the genomic region of streptococcal pyogenic
exotoxins. Davies et al.\ \citet{Davies2007} investigated phage 3396 from SDE as a vehicle of
gene transfer from SDE to SPY. They \citet{Davies2009} also demonstrated that an
integrative conjugative element of SDE allowed gene transfer from SDE to other
streptococci. Ahmad et al.\ \citet{Ahmad2009} did not conclusively deduce the net
directionality of housekeeping gene transfer between SPY and SDE.  Notably,
Kalia et al.\ \citet{Kalia2001} attempted to show that the net direction of gene transfer from
SPY to SDE was more dominant than the other direction.  Because these researches
on the controversial issue, net directionality of gene transfer between SPY and
SDE, have been limited to relatively small regions in the genomes primarily on
representative loci from the Multi Locus Sequence Typing (MLST) database, it
remains to be elusive to determine the genome-wide degrees to which genomic
segments flowed between the two species, and net directions of thereof.

As sequencing costs plummet \citep{Mardis2011}, bacterial genome sequencing
produces individual genomes of high quality even within a bacterial species
\citep{Tettelin2009a}.  Studies on horizontal gene transfer in prokaryotes have
used genes in very large sets of species from a great deal of genomes available 
\citep{Koonin2001}.  Although this type of study gave us insights into deep
evolutionary histories of prokaryotes, with multiple individual genomes per
species we now can go beyond such rough broad study on gene transfer towards
fine-scaled history of gene transfer between very closely related prokaryotes.
For instance, Liu et al.\ \citet{Liu2009} studied gene transfer events in
\textit{Lactobacillus bulgaricus} and \textit{Streptococcus thermophilus} using
the compositional approach of Karlin \citet{Karlin2001}. Luo et al.\ \citet{Luo2011} investigated
gene transfer events in \textit{Eschericha Coli} using the embedded quartet
decomposition due to Zhaxybayeva et al.\ \citet{Zhaxybayeva2006}. Hamady et al.\ \citet{Hamady2006} combined
phylogenetic and compositional approaches to detecting horizontal gene transfer.
Caro-Quintero et al.\ \citet{Caro-Quintero2011} studied a number of \textit{Shewanella baltica}
genomes to identify the genetic elements that enabled the species to adapt to
redox gradients.  Although several complete genomes from SPY have been
available, Suzuki et al.\ \citet{Suzuki2011} just recently added one more complete genome of
SDE to the set of a previously unique SDE genome \citep{Shimomura2011}. These
multiple SDE genomes prompted us to study genome-wide gene flows between SPY and
SDE that afflicted human beings around the globe.

Bacteria evolves new biological features through horizontally transferred genes
from other lineages and vertically transmitted mutations on genes within the
same lineage.  While mutations on a gene modifiy relatively small portions of the 
gene through substitutions, insertions, and deletions of deoxyribonucleic acids,
gene transfers can reshape the organization of a bacterial genome,
which could revamp biological functions of the species.  Recipient bacteria of
transferred genes can substitute transferred genes for homologous parts of their
genomes, or integrate transferred genes into their genomes without replacement.
These two processes are here in this manuscript discriminated by calling the
former \textit{recombining gene transfer}, and the latter \textit{horizontal
gene transfer} \citep{Ochman2001,Lawrence2009}.  Despite the two distinct
evolutionary processes in bacteria, analyses of complete genome sequences in
efforts of better understanding bacterial evolution have paid less attention to
a combined approach to illuminating disparity of the two gene transfer
processes. We attempted here to use and compare two approaches for
distinguishing the recombining gene transfer process from the horizontal, or
non-recombining, gene transfer, and tackled the two processes in association with
virulence genes.

Recently, a method of inferring homologous recombination with whole bacterial
genomes was developed \citep{Didelot2010}. The model of homologous recombination
allowed them to study recombining gene flows between bacteria individuals of
\textit{Bacillus cereus} species in the core genomes.   In the model, a species
tree that relates individuals genomes is augmented by attaching
\textit{recombinant edges} between species tree branches that can be used to
measure gene transfers.  Other evolutionary processes acting on bacterial
genomes include genome rearragement, translocation, inversion, duplication,
gain, loss, etc.  These processes did not readily lend themselves to
statistically more sound model-based approaches such as maximum likelihood.
Instead, a parsimony-based approach to inferring gene duplication, loss, and
transfer events was developed by Doyon et al.\ \citet{Doyon2011}.  In the
approach, gene trees are inferred for gene families that are constructed from
all of the genomes, and are reconciled with a speices tree to infer
parsimoniously events.  We noted that the first approach could be used to infer
transfer of genomic segments due to homologous recombination, and the second
could be useful in studying horizontal gene transfer.

In short, the complete genomes of the two species of SPY and SDE presented us a
unique opportunity to study gene transfer over relatively recent time scales.
We took advantage of the model-based approach for inferring homologous
recombination, and the parsimony-based one for detecting gene duplication, loss,
and horizontal transfer.  We found that virulence genes were more associated
non-recombining horizontal gene transfer than non-virulence genes.  We compared
both of the results in recombining gene transfer events to show that the
parsimony-based method inferred recombining gene transfer events in biological
functional categories similar to those inferred with the model-based approach.
This was the first attempt as far as we know of systematically analyzing SPY and
SDE in genome-wide scale to detect gene transfer events not only in the core
genome but also in the pan genome, which made the analysis complete in terms of
the amount of used genome data. Our analysis of the core genomes also showed for
the first time that the net directionality of recombining gene transfer from SPY
to SDE was stronger at the genome scale.

% Results and Discussion can be combined.
\section*{Results}

% \texttt{We show a net directionality of recombining gene transfers from SPY to SDE. 
% Describe our arguments of the net directionality. 
% Number of recombinant edges.
% Confidence about the numbers of recombinant edges.
% Not all of genes experience recombining gene transfers from SPY to SDE.
% Gene transfer counts from the parsimony-based approach; the power of the
% parsimony-based approach may be limited.
% Is this number of recombinant edges good enough for the argument?
% Reflection of the previous analyses, 
% heat maps of real data analysis, 
% and supporting simulations.}

\subsection*{Recombining gene transfer from SPY to SDE at genome-wide scale}

We followed the pipeline and methods recommended by Didelot et al.
\cite{Didelot2010} to pursue a model-based study of recombining gene transfer. A
preliminary step in the pipeline involved estimating the species tree with
branch lengths via \texttt{ClonalFrame} \cite{Didelot2007}, and three population
parameters of mutation rate, recombination rate, and average recombinant tract
length via a first phase of \texttt{ClonalOrigin} \cite{Didelot2010}.  Because
\textit{S.\ equi} ssp.\textit{\ equi} strain 4047 (SEE) diverged significantly
from the others among the six genomes (Table S\ref{tab:genome}), we could not
obtain sufficiently long multiple sequence alignments with it, and had to
exclude it from the model-based analysis.  The estimated species tree using the
alignment of the five genomes (Figure S\ref{fig:mauve}) had the expected
topology \texttt{SC: The figure 1 must be nicer than one that now} (see Figure
\ref{fig:tree5}). \texttt{ClonalFrame} also estimated a relative strength of
recombination compared to mutations as $r/m=4.66$ meaning that four times more
sites had been affected by recombination-driven substitutions than by
mutation-driven substitutions (Table S\ref{tab:clonalframe}).  This indicated
potentially significant roles of recombination in the evolutionary process of
shaping these genomes.  Estimates of population mutation rate ($\theta$),
recombination rate ($\rho$), and average tract length ($\delta$) were obtained
for the alignment blocks analyzed (Figure S\ref{fig:scatter3}), and a median of
the estimates weighted over block lengths was obtained as a global estimate for
each of the three parameters (Table S\ref{tab:three}).  The local estimates
typically seemed to be moderately fluctuating around the global estimate, with
the exception of what appeared to be a recombination hotspot at a region 200
kilobase pairs from the origin of replication of SDE1.  

The estimated species tree together with the global estimates of the three
population parameters were used to sample recombinant trees from the posterior
distribution in the second phase of \texttt{ClonalOrigin} \cite{Didelot2010}.
We summarized the posterior sample of recombinant trees by counting topologies
of the local genealogies induced by those recombinant trees for each of the
genomic positions (Table S\ref{tab:Gene-tree-topologies}).  Of the 105 possible
rooted topologies relatig five taxa, the dominant topology (66.7\%) was that of
the species tree.  The second most frequently sampled topology (9.3\%) was the
one in which SPY and SDE were sister clades to each other with SDD being as an
outgroup. Other topologies were sampled with frequency less than 4\%. This gave
initial indication to the extent of recombining gene transfer  between SPY and
SDE.

To gain a fine-scale view of the main types of recombination scenarios, we
looked at recombinant edges present in the posterior sample of recombinant
trees. Following the approach of Didelot et al. \cite{Didelot2010}, we
classified these edges by types defined by distinct pairs of donor and recipient
branches of the species tree (Figure \ref{fig:tree5}). For each edge type, we
averaged the number of edges present in the sampled recombinant tree, and
computed the ratio of this observed average frequency to a prior expected
frequency. The prior expected frequencies was computed by considering branch
lengths and their relative positions of the species tree, and the three
population parameter estimates in the preliminary step.  Base 2 logarithms of
these ratios are depicted in the heatmap of Figure
\ref{fig:Heatmap-of-recombination} (see also Tables S\ref{tab:heatmap} and
S\ref{tab:obsheatmap}).

Among the possible recombinant edge types, we can distinguish between ones that
preserve the species tree topology, and ones that alter it. We focused on 
topology-altering edge types both because our main objective was to study
cross-clade gene transfer, and because our simulation study indicated that the
observed counts of topology-preserving recombinant edges were unreliably
inferred (see \textbf{Simulations}). The heatmap allowed to find recombination
scenarios that deviated from the expectation dictated by the probabilistic
model.  We envisioned two main potential causes for such deviations: (1) reduced
or increased opportunities for species to interact and exchange genetic
material; (2) selection acting on newly transferred genes. 

We observed reduction in the number of recombinant edges sampled between SDD and
the three branches of the SPY clade ($\sim$25\% of the expected count). This was
consistent with the fact that these species had few opportunities to interact
with each other, SPY being a human pathogen and SDD being a strictly veterinary
one.  On the contrary, the number of sampled recombinant edges between SDD and
branches in the SDE clade seemed to be consistent with the prior.  The greatest
deviation of observed counts from expected ones was in recombinant edges between
SPY and SDE clades. It was concordant with the fact that these two species
shared an ecological niche; SDE was an opportunistic human pathogen.
Interestingly, there appeared to be much higher enrichment (up to four times
more than expected) for recombinant edges originating in the SPY clade and
ending up in the SDE clade.  In a follow-up simulation study we concluded that
this deviation from the expected count was indeed highly significant, and so was
the observed asymmetry in pattern of recombination (see \textbf{Simulations}).
This net directionality of recombination pattern from SPY to SDE was especially
intriguing in light of the hypothesis that SDE acquired human pathogenicity via
gene transfer from SPY.

\subsection*{Parsimony-based reconciliation of gene trees with the species tree}

Figure \ref{fig:Gene-duplication-loss} shows inferred numbers of gene gain,
loss, duplication, and two different kinds of gene transfer events along
lineages of the tree that relates the six genomes. SPY had less genes than the
other species. All of the six genomes were used to construct 2314 gene families.
The largest family consisted of 59 genes while nearly half of gene families
(1066) had exactly one gene per species, or six genes total (Figure
S\ref{fig:famsizes}).  We estimated a larger number of recombining gene transfer
events than that of horizontal gene transfer events; 944 and 294, respectively.
Duplications in extant lineages were more frequent than those in the other
lineages.  SDE lineages from the root appeared to have increased in the number
of genes while the common ancestor of SPY1 and SPY2 decreased in the number of
genes.  Whereas gene loss events were either comparable to or less than gene
gain events along most of the lineages, it was not the case along the two internal
branches of SPY and SDE.  This indicated that SPY and SDE lineages had reduced
number of genes to specialize in their specific niches after each divergence
from their corresponding direct ancestors.

As in the model-based analysis, we attempted to see which type of horizontal or
recombining transfer events was most common in our inference.  We classified
events into types defined by donor-recipient branches in the species tree. We
recorded the number of inferred events of each type (Figures
S\ref{fig:hgt-heatmap} and S\ref{fig:mowgli-recomb-heatmap}). Unlike in the
model-based analysis, we could not normalize these counts by their expectation
because we did not have any model-based expectation for them, which restricted
our ability to interpret these counts. We did notice, however, a large number of
inferred recombining gene transfer events from the ancestral SPY branch to the
ancestral SDE branch, consistent with our observations from the model-based
analysis. Yet, we did not observe any pattern of a net directional horizontal
gene transfer while we observed a large number of horizontal transfers from the
ancesral SDE branch to SDD. 

\subsection*{Gene categories associated with recombining and horizontal gene transfer}

We set out to seek categories of genes that were significantly enriched for
recombining or horizontal transfer events. For horizontal transfer, we used a
score based on a number of events inferred by the parsimony-based method, and
for recombining gene transfer, we used both the parsimony-based inferred counts
and a recombination intensity score based on the model-based analysis (see
\textbf{Materials and Methods}). First, we checked association of transfers with
virulence genes, which were provided by a previous study of Suzuki et al.\
\citet{Suzuki2011}.  The set of virulence genes was found to be significantly
associated with horizontal gene transfers (P-value $<1.33 \times 10^{-11}$), but
not with recombining gene transfers. We also found a marginally significant
association between virulence genes and gene duplications (P-value $<7.13 \times
10^{-3}$).

We conducted similar tests for other functional categories of genes. Several
functional categories that were found to be associated with recombination
intensity were related with either membrane-bound molecules or aminoacyl tRNA
synthetases (Table \ref{tab:functional}). Consistent results were obtained for
the parsimony-based counts of recombining gene transfers (Table
S\ref{tab:go-events-recombining}).  Functional categories associated with
horizontal gene transfer were clearly different from those associated with
recombining gene transfer (Table \ref{tab:go-events}). The top ranked functional
categories in that list were related with transposases. These differences added
to our previous findings about the different roles the two types of gene
transfer had played in the evolution of these species.

\subsection*{Comparison of model-based and parsimony-based inferences of recombining gene transfer}

The similarity between the two lists of gene categories found to be associated
with recombining gene transfers in the separate analyses -- model-based and
parsimony-based --  indirectly indicated concordance between the two independent
approaches.  In order to obtain a more direct measure of correlation, we used
the set of genes present in both analyses to compute the correlation between
recombination intensity scores and the number of recombining gene transfers
inferred by the parsimony-based approach. We observed a positive correlation of
$0.39$ (P-value $2.2\times10^{-16}$; Figure S\ref{fig:cmpcomowgli}) between the
two measures.  For comparison, the correlation between recombination intensity
and the number of horizontal gene transfers was 0.03 (P-value 0.18;
Figure S\ref{fig:cmpcomowglihgt}).  

\subsection*{A closer study of virulence genes}

Often computational and statistical analyses on genomes could benefit other
researchers when those resources were available in a form of genome browser.  We wished to
provide research community especially in SPY and SDE bacteria with results from
our analysis of gene transfer.  The resource shown in Figure S\ref{fig:ucsc} for
gene transfer was available at \url{http://strep-genome.bscb.cornell.edu}.  For
each of the five individual genomes as a reference genome several annotation
tracks were displayed: alignment blocks, recombination rate per site for each
block, local gene tree topologies, recombination intensity, locations of
virulence genes, posterior probability of recombining gene transfer. The genome
track of probability of recombining gene transfer allowed us to easily browse
the bacterial genome to locate regions that we were interested in.  Using
the genome browser tracks we located virulence genes with notable posterior
recombination probability (Table S\ref{tab:virrecomb}).  About 10\% of the genes
comprised of virulence facters, about half of which were found in the core
genome of the five genome alignment \citep{Suzuki2011}.  After sorting virulence
genes by the number of horizontal gene transfer events we listed several genes
with high ranks (Table S\ref{tab:virhgt}).
The whole gene of  putative dipeptidase (SpyM3\_0465) was transferred, which was
a exemplary case for the finding \citep{Chan2009} that virulence genes were transferred as a whole
unit.  Another example of whole gene transfer was superoxide
dismutase (SpyM3\_1071).  We investigated the unit of recombining gene transfer
by finding 58 genes for which a recombinant edge of certain types were sampled
with probability at least 0.9 by focusing on specific types of recombinant edges
(Table S\ref{tab:genes-transfer}).  In some cases, parts of a gene were inferred
to have been transferred, for instance, transfer from SPY1 to SDE1 was inferred
for 29\% of the aspartyl-tRNA synthetase gene. Recombinant tracts often were
shown to span two genes.  In other cases, the inferred transfer units spaned a
set of as much as eight adjacent genes: loci from SpyM3\_1518 to SpyM3\_1525.

\subsection*{Simulation study and additional support for biased directionality
in recombining gene transfer between SPY and SDE}

We conducted a series of experiments on simulated data to check the reliability
of estimates obtained by the model-based approach. We found that the population
parameters obtained in the first phase were reliable, with the exception that
the average recombinant tract length was over-estimated (this was observed also
in \citep{Didelot2010}). Our experiments also showed that the recombination
intensity used in our functional category association tests had fairly good
correlation between the true value and the simulated ones.  (Figure
S\ref{fig:ri1}). See \textbf{Materials and Methods} for a more detailed report
of these results. One thing we specifically wished to test in these experiments
was whether the model-based approach reliably detected deviations of recombinant
edge counts from their expectation, as observed in the heatmap of Figure
\ref{fig:Heatmap-of-recombination}. Our major concern was that these deviations
were not directly accounted for by the probabilistic model of ClonalOrigin, but
rather indirectly observed by summary of the posterior sample of recombinant
trees. 

We simulated the model to generate a hundred simulated data sets using a
collection of recombinant trees sampled from the prior distribution defined by
the species tree and three population parameters, as estimated from our genomic
data.  We then repeat the inference procedure of the second stage of
\texttt{ClonalOrigin} on each data set and, as in the data analysis, and counted
for each donor-recepient branch pair the average number of recombinant edges.
By normalizing the counts by the prior expected count we obtained intervals of
base 2 logarithms of the ratio of the observed number of recombinant edges to
the prior expectionation (Figure \ref{fig:h3}). The comparison showed a good
concordance of the estimated counts and the prior expected counts for all
topology-altering edge types and a clear over-estimation for the counts of
topology-preserving edges. Recall that we made sure to ignore these edges in our
analysis. For each edge type, we also plotted in Figure \ref{fig:h3} the
log-ratio of the count estimated for it in our data analysis and the expected
count (the closed circles without interval bars); these are the values making up
the heatmap of Figure \ref{fig:Heatmap-of-recombination}. Comparing the counts
estimated in the data analysis and the counts estimated from the simulated data
allowed to evaluate the significance of the divergence from expectation observed
in the heatmap.  Recall the heatmap indicated an excess of recombinant edges
between the SPY and SDE clades in both directions. Our simulations indicated
that the excess observed in the direction SDE to SPY was much less significant,
as the counts estimated from real data often falled very close to the
corresponding counts estimated from simulated data. The excess observed in the
direction SPY to SDE appeared not only to be larger, but also more significant.
This further supported our claim for the net directionality in recombining gene
transfer between these two species.

We wished to use simulated data to test not only the ability of the ClonalOrigin
sample to \textit{detect} deviations from expected counts of recombinant edges, but
also its ability to reliably \textit{quantify} this deviation. For this purpose we
generated a hundred simulated data sets using a collection of recombinant trees
sampled from the posterior sample of ClonalOrigin on the genomic data. The
difference between the two simulated batches was that in the first one, the
recombinant trees used represent the prior distribution, and in the second one,
they represented the posterior distribution, which we knew deviated from the
prior. Again, we recorded the recombinant edge counts obtained by ClonalOrigin
and normalized these counts by the prior expected counts (intervals with closed
circles in in Figure \ref{fig:h3}). We expected these counts to be in better agreement with
the counts observed in the data analysis. For recombinant edges that showed a
defficiency in our data analysis (e.g., from the SPY clade to SDD), the counts
appeared to fit well with the count estimated from genomic data. However, for
recombinant edges that showed an excess in our data analysis (e.g., from the SPY
clade to the SDE clade), the counts appeared to be significantly influenced by the
prior. We noted that the influence of the prior appeared to be even stronger for
topology-preserving recombinant edges. All things considered, we found that defficiency of
a certain recombinant edge type was reliably estimated whereas excesses were 
typically underestimated. Therefore, we concluded that the actual extent to which
recombining transfers between SPY and SDE \textit{preferred} the direction from SPY to
SDE was possibly somewhat underestimated in our analysis.

\section*{Discussion}

Molecular evolution of bacterial genomes is multifaceted. Researches of
bacterial evolution often focused on one side of many evolutionary processes.
While understanding one aspect of bacterial evolutionary processes we could be
ignorant of the other features.  Horizontal gene transfer process was often
considered in macro-evolutionary scale using a great deal of bacterial species.
As closely related species or groups of individual bacteria are becoming a focus
of evolutionary study, recombining gene transfer process would surface as
another facet of evolutionary forces.  Because the two gene transfer processes
would act on closely related bacterial genomes simultaneously, ignoring one of
the two processes in studying evolution of bacteria would lead to a narrowed
point of view to the underlying evolutionary forces. Our study of SPY and SDE
genomes undertook the work of discriminating recombining and horizontal gene
transfer processes because the two species were very closely related.  In this
study, we showed that the two gene transfer processes acted on different
biological functions.  Recombining gene transfer process in the analysis of SPY
and SDE genomes was related with genes encoding membrane-bound molecules and
aminoacyl tRNA synthetases. On the contrary, horizontal gene transfer process
was found to be associated with transposases, which would allow horizontal gene
transfer of their neighboring genes.  Because SPY was a human-specific pathogen,
and SDE was an opportunistic human pathogen, we were interested in association
of pathogenicity with the two gene transfer processes. We also showed that
virulence genes were more associated with horizontal gene transfer process than
with recombining gene transfer process.  With the current data of only SPY and
SDE genomes we could not conclude whether the association of virulence genes
with horizontal gene transfer not recombining gene transfer was applicable
generally to cases of other species.  However, without the discrimination of the
two processes it would have been difficult to find the disparate association of
biological functions or virulence genes with the different gene transfer
processes. 

We showed differences in the profile of functional category association of the
two gene transfer processes thanks to the approach of dicriminating the two gene
transfer processes.  Sugar phosphotransferases could be important in signal
transduction to recognize environmental changes.  Transporters, and fatty acid
and lipid biosyntheses would be involved in membrane-bound molecules.  Because
bacteria as a unicellular organism should be able to promptly respond to changes
to environment, some evolutionary pressures must have been acting on genes
encoding membrane-bound molecules.  We also found that the recombination
intensity and requency of recombining gene transfer events would be useful in
associating several aminoacyl tRNA synthetases genes belonging to gene ontology
term of translation (GO:0006412) with recombining gene transfer. It had been a
focus of gene transfer in bacterial species \citep{Woese2000}.  The biological
role of transposases was concordant with horizontal gene transfer process
because the process involved in transposases would allow bacteria to acquire
foreign genes without the requirement of homologous regions in the bacteria.  

Another contribution of our study to bacterial gene transfer research is to
elucidate the net directionality of gene flow from SPY to SDE in the core
genomes. Although the issue of net directionality has been tackled, our study
was the first attempt of resolving the issue in genome-wide scale. Yet, it still
remains to be seen whether this net directionality of gene flow from SPY to SDE
happened in the pan genome. A model-based method with more refined models would
be necessary because the parsimony-based method we employed simply could not
show any net directionality between SPY and SDE in the pan genome. Although we
are unsure whether much larger number of genomes could be useful in finding net
directionality, more SDE genomes might be beneficial to future studies. 

Were there unbalanced recombining gene transfer between the two species, what
were the barriers \citep{Thomas2005} to cause the unbalanced gene transfer in
the species?  Kalia et al.\ \citet{Kalia2001} discussed restriction-modification
system (RMS) as a possible mechanism of the unbalanced gene flow.
Restriction-modification system is a biological self-defense mechanism of
bacteria, which is diverse in prokaryote world.  As biological roles of RMS
might be to maintain and control species identity \citep{Jeltsch2003}, Budroni
et al.\ \citet{Budroni2011a} recently studied that \textit{Neisseria
meningitidis} phylogenetic clades were associated with RMS, claiming that RMS
modulated homologous recombination in the bacteria of \textit{N.\ meningitidis}.
Most of \textit{S.\ pyogenes} genomes contained restriction-modification type I
system as well as type II system. The two genomes of SDE contained only type II
system. More tight self-defense system of SPY may have made the species less
capable of aquiring foreign genetic material from SDE \citep{Kalia2001}.

We reexamined the seven genes that Kalia et al.\ \citet{Kalia2001} had studied
for net directionality of gene transfer between SPY and SDE.  The internal
fragments of the six housekeeping genes (\textit{gki}, \textit{gtr},
\textit{murI}, \textit{mutS}, \textit{recP}, \textit{xpt}) were located in the
core genome that we investigated in this study.  Except for \textit{mutS} where
Kalia et al.\ \citet{Kalia2001} did not find recombination, our results were
consistent with those of Kalia et al.\ \citet{Kalia2001}.  Pinho et al.\
\citet{Pinho2010} reported recombination in the gene \textit{parC}, in which we
also observed moderate recombination.

It remains to be seen whether the inability of detecting net directionality
between SPY and SDE in the pan genomes was because of lack of power of the
parsimony-based approach that we employed. Considering gene duplication, loss,
and horizontal gene transfer in a model-based approach would be desirable in the
coming ages of abundant bacterial genomes. The model might consider genome
alignments where each taxon does not necessarily have a unique gene; some taxa
might have no gene due to gene loss, and other taxa multiple genes due to gene
gain, duplication, or horizontal gene transfer. With these types of models will
we be able to address more interesting evolutionary questions in bacteria. This
need of more refined models would be also true even for the study of net
directionality between species in the core genome.  In other words, one could
envision a model where a population two-taxa (i.e., SPY and SD groups) tree is
superimposed on the species tree shown in Figure S\ref{fig:clonalorigin}A. A
splitting event of the population tree could happen somewhere between the root
of the species tree and the internal node of SPY. Recombinant edges that connect
branches within SPY or SD clades would be more frequent than those that connect
branches between SPY and SDE clades.  With the reference to the recombination
within a clade one parameter can dictate the degree of gene from from SPY to
SDE, and another for the reverse direction.  If the two parameters could be
incorporated in the model, then they might be used to compare the degrees of
gene flow between the two species in a more statistically sound way.  We hope
that we will see this refinement of the model to better illustrate bacterial
evolution in forthcoming bacterial genomic era.

% You may title this section "Methods" or "Models". 
% "Models" is not a valid title for PLoS ONE authors. However, PLoS ONE
% authors may use "Analysis" 
\section*{Materials and Methods}

\subsection*{Complete bacterial genomes}

Two genomes of \textit{S.\ pyogenes} (SPY), and two of \textit{S.\ dysgalactiae}
ssp.\textit{\ equisimilis} (SDE) were downloaded from NCBI with the following
accessions: NC\_004070 \citep{Beres2002}, NC\_008024 \citep{Beres2006},
NC\_012891 \citep{Shimomura2011}, and CP002215 \citep{Suzuki2011}. We refer to
these samples as SPY1, SPY2, SDE1, and SDE2, respectively. The genome of
\textit{S.\ dysgalactiae} ssp.\textit{\ dysgalactiae} was provided by
Suzuki et al.\ \citet{Suzuki2011} with accession identifiers of SddyATCC27957, referred to here
as SDD. We also downloaded the genome (NC\_012471) of \textit{S.\ equi}
ssp.\textit{\ equi} strain 4047 referred to as SEE \citep{Holden2009}.

\subsection*{Model-based method for recombining gene transfer}

We present a short summary of the statistical method developed by
Didelot et al.\ \citet{Didelot2007} and \citet{Didelot2010} for studying bacterial
recombination. A detailed account was found in the two papers and in the user
guide available at \url{http://code.google.com/p/clonalorigin}.  We used
\texttt{progressiveMauve} v2.3.1 \citep{Darling2004,Darling2010} with default
options to attempt to align the six genomes, which led to too small proportion
of core genomes. We had to remove the SEE sample from the genome alignment
although we kept it in the parsimony-based approach.  We then used
\texttt{stripSubsetLCBs}, a program of the \texttt{ClonalOrigin} package, and
identified 276 blocks of sequence alignments with length threshold of 1500 base
pairs.  We excluded two of these blocks in a subsequent analysis because they
contained long consecutive gaps.  We applied \texttt{ClonalFrame} v1.1 available
at \url{http://www.xavierdidelot.xtreemhost.com} to the remaining 274 blocks,
and estimated the species tree relating the five genomes. \texttt{ClonalFrame}
also estimates the fraction between recombination and mutation rates.
We used a single Markov chain Monte Carlo (MCMC) with 10000 burn-in generations
and 10000 sampling generations to obtain the tree estimate with branch lengths,
and additionally seven independent MCMC runs to validate adequate convergence of
the first run.

Two stages of MCMC with \texttt{ClonalOrigin}, subversion r19 that had been
downloaded from \url{http://clonalorigin.googlecode.com/svn/trunk}, were
employed independently for each of the 274 alignment blocks as described in
Didelot et al.\ \citet{Didelot2010}.   In the first MCMC we obtained estimates of mutation rate,
recombination rate, and recombinant tract length.  Two independent chains were
executed with $10^6$ burn-in generations and $10^7$ sampling generations. We
sub-sampled the result every $10^5$-th generations.  A few alignment blocks were
not successfully finished even after a month of computation.  Because those were
not expected to affect the global estimates of population parameters
significantly, we used only finished blocks to estimate mutation rate,
recombination rate, and recombinant tract length.  The global estimators of the
three population parameters were medians weighted over block lengths, which
down-weighed less reliable estimates obtained from shorter alignment blocks.
One of the two independent chains was used to obtain the block-wise mean
estimates, and the other chain was used to validate convergence of the first
stage of MCMC.  Two independent MCMC runs resulted in similar estimates of the
three population parameters (Table S\ref{tab:three}).  The second stage of MCMC
with \texttt{ClonalOrigin} was executed with the three population parameters
fixed to the estimated values in the first stage.  Two independent MCMC chains
were also executed with a total generation being $1.1\times10^8$. The first
$10^7$ generations of each chain were discarded as burn-in. The remaining $10^8$
generations were sub-sampled every $10^5$-th generation.  

\subsection*{Parsimony-based method for horizontal gene transfer}

While \texttt{ClonalOrigin} could be used to understand recombining gene
transfers within the conserved core genome, we sought to understand the history
of the remainder of the genome where more complex evolutionary processes occur,
such as gene duplications, losses, and horizontal transfers.  In recent years,
several methods have been developed for inferring these evolutionary events in
gene families by reconciling gene trees with a species tree
\citep{David2011,Doyon2011,Tofigh2011}.  In this analysis, we complemented our
study of recombination by using \texttt{Mowgli} \citep{Doyon2011} to
characterize events that can change the copy number of a gene, such as
duplications, losses, and non-recombining horizontal gene transfers.  In
contrast to \texttt{ClonalOrigin}, \texttt{Mowgli} is parsimony-based for
inferring events and uses pre-determined gene clusters where genes are treated
as atomic units.  The advantage of using \texttt{Mowgli} was that it did not
assume the sequence data contains a single copy of each homologous segment, and
it attempted to explain different copy numbers by a series of gain and loss
events.  We obtained gene family clusters defined in Suzuki et al.\ \citet{Suzuki2011} using
the \texttt{OrthoMCL} method \citep{Li2003} over the full set of the six genomes
including \textit{S.\ equi} subsp.\textit{\ equi} strain 4047 (SEE).  For each
family cluster, the corresponding protein sequences were aligned using the
\texttt{MUSCLE} program \citep{Edgar2004a} and nucleotide alignments were
constructed by mapping nucleotide sequences onto each protein sequence
alignment. We then identified recombination within gene alignments using the
Single Breakpoint Recombination (\texttt{SBP}) method from the \texttt{HyPhy}
package \citep{KosakovskyPond2006}.  Using the Akaike Information Criterion
(AIC), \texttt{SBP} inferred that 47.5\% (1,094) of gene families contained at
least one topology-changing recombination. Each of these families was then split
into two gene fragments.  Gene trees were constructed from nucleotide alignments
for each gene fragment using the \texttt{RAxML} program \citep{Stamatakis2006},
and were rooted by minimizing inferred duplications and losses.  Each rooted
gene tree was then reconciled using the \texttt{Mowgli} program
\citep{Doyon2011}, which inferred the most parsimonious gene duplication, loss,
and transfer events for each family.  In the case of multiple maximum
parsimonious reconciliations, one was chosen uniformly at random.  Unless
specified otherwise, default setings were used for all the programs.  Since the
method of \texttt{Mowgli} did not directly model recombination events,
recombination-based transfers were inferred in a post-processing step, by
checking whether the transfer was coupled with a loss of a homologous gene in
the recipient branch (indicating gene replacement rather than gain).  A transfer
event was classified as non-recombining if the destination species of that
transfer had an extant descendant that contained copies of that genes which were
descendants of the transferred gene as well as copies which were
non-descendants.  We noted that this parsimonious approach was imperfect and
would tend to undercount events, especially recombining gene transfer events,
which were modeled using two events: a transfer and a loss.  Nevertheless, this
analysis allowed us to study non-recombining gene transfers, which were not
modeled by \texttt{ClonalOrigin}, and to validate some of the
\texttt{ClonalOrigin} results on recombining gene transfer.  

\subsection*{Functional categories association}

We performed a series of analyses to check for functional gene categories
enriched for certain types of gene transfer events. Functional categories were
assigned in the same manner as Suzuki et al.\ \citet{Suzuki2011}. In short, \texttt{blastP} was
used to compare Streptococcus genes to bacterial proteins from the Uniref90
database.  Matches with E-values less than $1.0\times10^{-5}$ were assumed to
have the same Gene Ontology (GO) classification as the target gene, as
determined from the uniProt GOA database.  We also classified a category of
\textit{virulent genes} using the virulence profile described in
Suzuki et al.\ \citet{Suzuki2011}. Among 207 virulent genes at least one of the genes in the
gene family was identified by that profile.  Each gene covered by our analysis
was associated with a recombination intensity score from the
\texttt{ClonalOrigin} analysis and scores from the \texttt{Mowgli} analysis
corresponding to number of duplications, losses, horizontal gene transfers, and
recombining gene transfer events.  Mann-Whitney tests were used to identify GO
categories whose genes showed significantly elevated values in each score.  A
5\% False Discovery Rate (FDR) correction \citep{Benjamini1995} was used to
determine an appropriate significance threshold for our p-values.

\subsection*{Simulation Study}

We performed three sets of simulations.  In the first simulation, we wished to
show that \texttt{ClonalOrigin} could infer population mutation rate, tract
length, and recombination rate with reasonable accuracy. In the second
simulation, we wished to check whether recombinant edges were recovered from
simulated data with recombinant trees generated from the prior. The third
simulation was conducted using simulated data with recombinant trees sampled the
posterior distribution.  We used a simulation feature of \texttt{ClonalOrigin}
to simulate the process of mutation and recombination across the five species
(see section \textbf{Supplementary Information} for detail).


% Do NOT remove this, even if you are not including acknowledgments
\section*{Acknowledgments}

We are indebted to Xavier Didelot for his help in setting up \texttt{ClonalOrigin}
analysis.  We are thankful to Haruo Suzuki for his help
in using the data of SDD genome. This work was supported by the National
Institute of Allergy and Infectious Disease, US National Institutes
of Health, under grant number AI073368-01A2 awarded to A.S. and M.J.S.

\section*{Author Contributions}

Conceived and designed the experiments: SCC MDR MJS AS.
Performed the experiments: SCC MDR MJH.
Analyzed the data: SCC MDR MJH IG.
Contributed reagents/materials/analysis tools: MJS.
Wrote the paper: SCC MDR MJH IG AS.

%\section*{References}
% The bibtex filename
\bibliographystyle{plos}
\bibliography{siepel-strep}
\clearpage{}

\section*{Figure Legends}

\begin{flushleft}

Figure \ref{fig:tree5}.\ {\bf The species tree of the five genomes.}
SPY is the branch of a clade of \textit{Streptococcus
pyogenes}, SDE is that for \textit{S.\ dysgalactiae} ssp.\textit{\ equisimilis},
and SD is the branch for the three individual genomes of \textit{S.\
dysgalactiae}. The scale is proportional to the expected number of mutations
given the Watterson's estimate of mutation rate and the tree.

Figure \ref{fig:Heatmap-of-recombination}.\ {\bf Heatmap of recombination events 
using the five genomes.}
Each cell of the heat map represents the logarithm base
2 of the number of recombination events inferred relative to its expectation
under the prior, for a donor and recipient pair of branches. The donor is given by
the y-axis; the recipient on the x-axis.  The black colored cells are meaningless
because the recipient species tree branch of the cells are older than the donor
species branch.

Figure \ref{fig:Gene-duplication-loss}.\ {\bf Distribution of inferred gene 
duplication, loss, and transfer events across
the phylogeny.}  Gene counts for
each extant and ancestral species are given at each node of the phylogeny.  For
each branch we denote the number of gains G*, duplications D*, losses L*,
non-recombining transfers T*, and recombinations R*.  The non-recombining
transfer and recombination counts are indicated on the recipient branch (donors
are not indicated).

Figure \ref{fig:h3}.\ {\bf Simulations of the model-based approach for assessing 
the power of recovering recombinant edges.} 
Pairs of species tree branches are shown at x-axis labels.
Three intervals are displayed for each pair. Intervals with solid line with open
circle in the middle are for the logarithm of the fraction to base 2 of the average
number of recombinant edges from the ten simulated recombinant trees 
to the numbers of recombinant edges estimated from Simulation 2. 
Interverals with dotted line are for the logarithm of the fraction to base 2 of
the number of recombinant edges estimated from the real data
to the numbers of recombinant edges estimated from Simulation 2.
Intervals with solid line with closed circle in the middle are for the logarithm
of the fraction to base 2 of the number of recombinant edges estimated from the real data
to the numbers of recombinant edges estimated from Simulation 3. 
The horizontal lines are at the zero.
The intervals range from 5\% quantile to 95\% of quantile.

\end{flushleft}
\clearpage{}

%\begin{figure}[!ht]
%\begin{center}
%%\includegraphics[width=4in]{figure_name.2.eps}
%\end{center}
%\caption{
%{\bf Bold the first sentence.}  Rest of figure 2  caption.  Caption 
%should be left justified, as specified by the options to the caption 
%package.
%}
%\label{Figure_label}
%\end{figure}

\section*{Tables}
%\begin{table}[!ht]
%\caption{
%\bf{Table title}}
%\begin{tabular}{|c|c|c|}
%table information
%\end{tabular}
%\begin{flushleft}Table caption
%\end{flushleft}
%\label{tab:label}
% \end{table}

%=============================================================================
% Table 1. 
\begin{table}[!ht]
\caption{
{\bf Functional categories associated with 
recombination intensity based on topology-altering recombinant edges.}}
\noindent \begin{centering}
\begin{tabular}{cccl}
\hline 
P-Value & Count & GO term & Description\tabularnewline
\hline 
1.731e-06 & 38 & GO:0009401 & phosphoenolpyruvate-dependent sugar phosphotransferase system\tabularnewline
2.775e-06 & 33 & GO:0008982 & protein-N(PI)-phosphohistidine-sugar phosphotransferase activity\tabularnewline
4.581e-05 & 22 & GO:0006633 & fatty acid biosynthetic process\tabularnewline
0.0001248 & 20 & GO:0008610 & lipid biosynthetic process\tabularnewline
0.0002675 & 45 & GO:0008643 & carbohydrate transport\tabularnewline
0.000513 & 81 & GO:0006412 & translation\tabularnewline
0.0005617 & 13 & GO:0006814 & sodium ion transport\tabularnewline
0.001019 & 40 & GO:0030529 & ribonucleoprotein complex\tabularnewline
0.001203 & 12 & GO:0031402 & sodium ion binding\tabularnewline
0.00206 & 39 & GO:0003735 & structural constituent of ribosome\tabularnewline
0.002115 & 237 & GO:0008152 & metabolic process\tabularnewline
0.00215 & 26 & GO:0019843 & rRNA binding\tabularnewline
0.00287 & 371 & GO:0016740 & transferase activity\tabularnewline
0.003552 & 66 & GO:0005840 & ribosome\tabularnewline
\hline 
\end{tabular}
\par\end{centering}
\label{tab:functional}
\end{table}
%\clearpage{}
%=============================================================================

%=============================================================================
% Table 2. 
\begin{table}[!ht]
\caption{
{\bf Functional categories associated with 
non-recombining horizontal gene transfer.}}
\noindent \begin{centering}
\begin{tabular}{cccl}
\hline 
P-Value & Count & GO term & Description \\
\hline 
7e-61 &  67 & GO:0006313 & transposition, DNA-mediated\\
3e-58 &  59 & GO:0004803 & transposase activity\\
1e-30 &  52 & GO:0015074 & DNA integration\\
1e-28 &  23 & GO:0032196 & transposition\\
1e-13 & 527 & GO:0003677 & DNA binding\\
2e-12 &  86 & GO:0006310 & DNA recombination\\
1e-11 & 197 & GO:0003676 & nucleic acid binding\\
3e-09 &  36 & GO:0016987 & sigma factor activity\\
3e-09 &  36 & GO:0006352 & transcription initiation\\
3e-08 & 142 & GO:0043565 & sequence-specific DNA binding\\
\hline 
\end{tabular}
\par\end{centering}
\label{tab:go-events}
\end{table}
\clearpage{}
%=============================================================================

\section*{Figures}

%=============================================================================
% Figure 1: Species tree
\begin{figure}[!ht]
\includegraphics[scale=0.7]{figures/cornellf-3-tree}
\caption{
{\bf The species tree of the five genomes.} 
SPY is the branch of a clade of \textit{Streptococcus
pyogenes}, SDE is that for \textit{S.\ dysgalactiae} ssp.\textit{\ equisimilis},
and SD is the branch for the three individual genomes of \textit{S.\
dysgalactiae}. The scale is proportional to the expected number of mutations
given the Watterson's estimate of mutation rate and the tree.}
\label{fig:tree5}
\end{figure}
\clearpage{}%
%=============================================================================

%=============================================================================
% Figure 2. Heat map of recombining gene transfer from ClonalOrigin
\begin{figure}
\includegraphics[scale=0.45]{figures/heatmap-recedge}
\caption{\label{fig:Heatmap-of-recombination}
{\bf Heatmap of recombination events using the five genomes.}
Each cell of the heat map represents the logarithm base
2 of the number of recombination events inferred relative to its expectation
under the prior, for a donor and recipient pair of branches. The donor is given by
the y-axis; the recipient on the x-axis.  The black colored cells are meaningless
because the recipient species tree branch of the cells are older than the donor
species branch.}
\end{figure}
\clearpage{}
%=============================================================================

%=============================================================================
% Figure 3. Gene duplication, loss, and transfer
\begin{figure}
\includegraphics[width=6in]{figures/strep-events}
\caption{\label{fig:Gene-duplication-loss} 
{\bf Distribution of inferred gene duplication, loss, and transfer events across
the phylogeny.}  Gene counts for
each extant and ancestral species are given at each node of the phylogeny.  For
each branch we denote the number of gains G*, duplications D*, losses L*,
non-recombining transfers T*, and recombinations R*.  The non-recombining
transfer and recombination counts are indicated on the recipient branch (donors
are not indicated).}
\end{figure}
\clearpage{}%
%=============================================================================

%=============================================================================
% Figure 4. 
\begin{figure}

\subfloat[]{\includegraphics[scale=0.4]{figures/h3-R-ratio-topology-changing}}

\subfloat[]{\includegraphics[scale=0.4]{figures/h3-R-ratio-not-topology-changing}}

\caption{
{\bf Simulations of the model-based approach for assessing the power of recovering
recombinant edges.} Pairs of species tree branches are shown at x-axis labels.
Three intervals are displayed for each pair. Intervals with solid line with open
circle in the middle are for the logarithm of the fraction to base 2 of the average
number of recombinant edges from the ten simulated recombinant trees 
to the numbers of recombinant edges estimated from Simulation 2. 
Interverals with dotted line are for the logarithm of the fraction to base 2 of
the number of recombinant edges estimated from the real data
to the numbers of recombinant edges estimated from Simulation 2.
Intervals with solid line with closed circle in the middle are for the logarithm
of the fraction to base 2 of the number of recombinant edges estimated from the real data
to the numbers of recombinant edges estimated from Simulation 3. 
The horizontal lines are at the zero.
The intervals range from 5\% quantile to 95\% of quantile.}
\label{fig:h3}
\end{figure}
\clearpage{}%
%=============================================================================

\clearpage{}\setcounter{figure}{0}
\setcounter{table}{0}
\renewcommand{\figurename}{Supplementary Figure}
\renewcommand{\tablename}{Supplementary Table}

\section*{Supplementary Tables}

%=============================================================================
% Table S1. List of six genomes.
\begin{table}[!ht]
\caption{
{\bf The six complete bacterial genomes.}
Each bacteria and its genome is referred to as the name at the first column.}
\noindent \begin{centering}
\begin{tabular}{cccc}
\hline 
Name & NCBI accession & Size (base pairs) & Reference\tabularnewline
\hline
SDE1 & CP002215 & 2159491 & Suzuki et al. (2011)\tabularnewline
SDE2 & NC\_012891 & 2106340 & Shimomura et al. (2011)\tabularnewline
SDD & N/A$^a$ & 2141837 & Suzuki et al. (2011)\tabularnewline
SPY1 & NC\_004070 & 1900521 & Beres et al. (2002)\tabularnewline
SPY2 & NC\_008024 & 1937111 & Beres et al. (2006)\tabularnewline
SEE & NC\_012471 & 2253793 & Holden et al. (2009)\tabularnewline
\hline
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$ SDD genome was not yet available from NCBI.
\end{flushleft}
\label{tab:genome}
\end{table}
%\clearpage{}
%=============================================================================

%=============================================================================
% Table S2. ClonalFrame Result.
\begin{table}[!ht]
\caption{
{\bf Parameter estimates from \texttt{ClonalFrame}.}}
\noindent \centering{}\begin{tabular}{cc}
\hline
Parameter & Estimates (95\% credibility interval) \tabularnewline
\hline
$\theta^a$ & 92871.84 (N/A)\tabularnewline
$\nu^b$ & 0.134 (0.132, 0.135)\tabularnewline
$R^c$ & 9370 (8950, 9780)\tabularnewline
$\delta^d$ & 362 (347, 377)\tabularnewline
$r/m^e$ & 4.66 (4.45, 4.87)\tabularnewline
$\rho/\theta^f$ & 0.101 (0.096, 0.105)\tabularnewline
TMRCA$^g$ & 0.331 (0.324, 0.337)\tabularnewline
$T^h$ & 1.02 (1.01, 1.04)\tabularnewline
\hline
\end{tabular}
\begin{flushleft}
$^a$ $\theta$ is population mutation rate,

$^d$ $\delta$ is the average recombinant tract length,

$^e$ $r/m$ is the relative strength of recombination compared to mutations,

$^g$ TMRCA is the time to most recent common ancestor, and

$^h$ $T$ is the total branch length of the inferred species tree.
\end{flushleft}
\label{tab:clonalframe}
\end{table}
%\clearpage{}
%=============================================================================

%=============================================================================
% Table S. ClonalOrigin's 1st stage result.
\begin{table}[!ht]
\caption{
{\bf Three population parameter estimates from the model-based approach.}
The estimates of the three parameters were based on one of two independent
MCMC chains.  We considered the similar 
results of the two independent chains as confirmation of 
the convergence of the first stage of MCMC.}
\noindent \begin{centering}
\begin{tabular}{ccc}
\hline
& Replicate 1 & Replicate 2\tabularnewline
Parameter & Estimate (IQR$^a$) & Estimate (IQR)\tabularnewline
\hline
$\theta^b$ & 0.081 (0.067, 0.094) & 0.081 (0.067, 0.099)\tabularnewline
$\rho^c$ & 0.012 (0.006, 0.019) & 0.012 (0.006, 0.020)\tabularnewline
$\delta^d$ & 744 (346, 2848) & 723 (348, 2870)\tabularnewline
\hline
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$ IQR stands for interquartile range.

$^b$ The median of mutation rate per site weighted over the lengths of blocks. 

$^c$ The median of recombination rate per site weighted over the lengths of blocks.

$^d$ The median value for the recombinant tract length weighted over the lengths of blocks.
\end{flushleft}
\label{tab:three}
\end{table}
%\clearpage{}
%=============================================================================

%=============================================================================
% Table S4. Gene tree topologies
\begin{table}[!ht]
\caption{
{\bf Posterior probability of gene tree topologies.}
See Table S\ref{tab:genome} for the reference of the bacteria
individual names.}
\noindent \begin{centering}
\begin{tabular}{cc}
\hline 
Gene tree topologies & Posterior probability\tabularnewline
\hline
(((SDE1,SDE2),SDD),(SPY1,SPY2)) & 66.7\%\tabularnewline
(((SDE1,SDE2),(SPY1,SPY2)),SDD) & 9.3\%\tabularnewline
((SDE1,SDE2),(SDD,(SPY1,SPY2))) & 3.5\%\tabularnewline
(((SDE1,(SDE2,SDD)),(SPY1,SPY2)) & 1.5\%\tabularnewline
(((SDE1,SDD),SDE2),(SPY1,SPY2)) & 1.3\%\tabularnewline
((SDE1,SDD),(SDE2,(SPY1,SPY2))) & 1.2\%\tabularnewline
((((SDE1,SDE2),SPY1),SPY2),SDD) & 1.0\%\tabularnewline
((SDE2,SDD),(SDE1,(SPY1,SPY2))) & 1.0\%\tabularnewline
\hline
\end{tabular}
\par\end{centering}
\label{tab:Gene-tree-topologies}
\end{table}
%\clearpage{}
%=============================================================================

%=============================================================================
% Table S4. The values of the heat map.
\begin{table}[!ht]
\caption{
{\bf Ratios of logarithm base 2 of the observed
recombinant edges to the expected number of recombinant edges.}  Figure
\ref{fig:Heatmap-of-recombination} shows the values graphically using heat map.
NA's are for the branch pairs that are impossible under the model-based
approached. Rows are for donor, and columns for recipient.}
\noindent \centering{}\begin{tabular}{cccccccccc}
\hline
& SDE1 & SDE & SDE2 & SD & SDD & ROOT & SPY1 & SPY & SPY2 \tabularnewline
\hline
SDE1& 0.176  &   NA  &1.170  &    NA & 0.7564  & NA & 0.67 &    NA & 0.663\tabularnewline
SDE & -0.118 & 0.081 &-0.202 &     NA&  0.3395 &  NA& -0.15&  0.045& -0.323\tabularnewline
SDE2& 1.201  &   NA  &0.092  &    NA & 0.6978  & NA & 0.60 &    NA & 0.387\tabularnewline
SD  & -0.629 & 0.022 &-0.661 &0.00670&  0.0078 &  NA& -0.94& -0.299& -1.030\tabularnewline
SDD & -0.317 & 0.371 &-0.197 &     NA&  0.0752 &  NA& -1.06& -0.525& -0.915\tabularnewline
ROOT& -1.745 &-0.698 &-1.786 &0.23039& -0.4854 &  NA& -1.37&  0.497& -1.306\tabularnewline
SPY1& 1.245  &1.883  &1.251  &    NA &-2.1997  & NA & 0.06 &    NA & 1.062\tabularnewline
SPY & 0.089  &0.200  &0.298 &0.00098 &-0.7903  & NA &-0.19 &-0.058 &-0.159\tabularnewline
SPY2& 1.174  &1.938  &1.287  &    NA &-1.9740  & NA & 1.07 &    NA & 0.063\tabularnewline
\hline
\end{tabular}
\label{tab:heatmap}
\end{table}
%\clearpage{}
%=============================================================================


%=============================================================================
% Table S4-1. The significance values of the heat map.
\begin{table}[!ht]
\caption{
{\bf P-values for the significance of the ratios of logarithm base 2 of the observed
recombinant edges to the expected number of recombinant edges.} 
NnN's are for the branch pairs that are impossible under the model-based
approached. Rows are for donor, and columns for recipient.}
\noindent \centering{}\begin{tabular}{cccccccccc}
\hline
&         SDE1&     SDE&    SDE2&      SD&     SDD &ROOT&    SPY1&     SPY&    SPY2\\
\hline
SDE1 &3.1e-03 &    NaN &1.1e-29 &    NaN &8.8e-21  &NaN &4.4e-19 &    NaN &2.2e-23\\
SDE  &4.9e-02 &2.3e-06 &7.5e-04 &    NaN &3.5e-32  &NaN &9.3e-01 &3.0e-08 &1.2e-25\\
SDE2 &8.2e-32 &    NaN &6.0e-01 &    NaN &1.2e-23  &NaN &1.6e-17 &    NaN &3.3e-06\\
SD   &6.4e-11 &8.0e-01 &1.4e-19 &3.2e-02 &7.3e-01  &NaN &2.2e-27 &1.5e-22 &1.1e-30\\
SDD  &7.3e-33 &6.7e-32 &4.1e-04 &    NaN &9.1e-07  &NaN &8.7e-34 &1.8e-32 &8.7e-34\\
ROOT &8.7e-34 &8.7e-34 &8.7e-34 &8.7e-34 &8.7e-34  &NaN &8.7e-34 &8.7e-34 &8.7e-34\\
SPY1 &8.7e-34 &2.9e-32 &8.7e-34 &    NaN &8.7e-34  &NaN &9.3e-01 &    NaN &8.7e-34\\
SPY  &2.6e-13 &2.2e-27 &5.8e-32 &7.5e-01 &8.7e-34  &NaN &1.3e-03 &3.9e-05 &6.5e-02\\
SPY2 &8.7e-34 &2.1e-33 &8.7e-34 &    NaN &8.7e-34  &NaN &9.4e-34 &    NaN &9.3e-01\\
\hline
\end{tabular}
\label{tab:heatmappvalue}
\end{table}

%=============================================================================
% Table S5. The values of the heat map of actual counts.
\begin{table}[!ht]
\caption{
{\bf The observed number of recombinant edges.}
The zeros are for the branch pairs that are impossible under the model-based
approached. Rows are for donor, and columns for recipient.}
\noindent \centering{}\begin{tabular}{cccccccccc}
\hline
& SDE1 & SDE & SDE2 & SD & SDD & ROOT & SPY1 & SPY & SPY2 \tabularnewline
\hline
SDE1&8.896 &  0.00 &17.712  & 0.00 & 12.872   & 0 &12.59  &  0.00  &12.53\tabularnewline
SDE &44.626& 136.49& 42.088 &  0.00& 227.061  &  0& 73.39 &  99.95 & 65.25\tabularnewline
SDE2&18.099&   0.00&  8.393 &  0.00&  12.360  &  0& 11.97 &   0.00 & 10.36\tabularnewline
SD  &9.671 &101.74 & 9.464  &31.26 &115.796   & 0 &13.68  & 98.38  &12.86\tabularnewline
SDD &45.539& 168.58& 49.513 &  0.00& 194.706  &  0& 43.09 &  67.14 & 47.65\tabularnewline
ROOT&43.553& 602.67& 42.337 &765.53& 802.813  &  0& 98.86 &2146.01 &103.60\tabularnewline
SPY1&40.800&  11.46& 40.973 &  0.00&   4.371  &  0& 21.15 &   0.00 & 42.35\tabularnewline
SPY &57.618& 262.08& 66.594 & 30.92& 163.518  &  0& 83.56 & 207.00 & 85.10\tabularnewline
SPY2&38.837&  11.90& 42.000 &  0.00&   5.111  &  0& 42.45 &   0.00 & 21.19\tabularnewline
\hline
\end{tabular}
\label{tab:obsheatmap}
\end{table}
%\clearpage{}
%=============================================================================


%=============================================================================
% Table S7. Mowgli - Recombining Transfer
\begin{table}[!ht]
\caption{
{\bf Functional categories enriched in 
recombining transfer events that are inferred using the parsimony-based
approach.}}
\noindent \begin{centering}
\begin{tabular}{cccl}
\hline 
P-value & Count & GO-term & Description \\
\hline 
7e-09 & 153 & GO:0006412 & translation\\
1e-08 &  67 & GO:0030529 & ribonucleoprotein complex\\
1e-07 &  45 & GO:0019843 & rRNA binding\\
2e-06 &  73 & GO:0003735 & structural constituent of ribosome\\
3e-05 & 157 & GO:0005840 & ribosome\\
7e-05 & 221 & GO:0003723 & RNA binding\\
2e-04 &  11 & GO:0004826 & phenylalanine-tRNA ligase activity\\
3e-04 &  14 & GO:0015413 & nickel-transporting ATPase activity\\
4e-04 &  93 & GO:0008982 & protein-N(PI)-phosphohistidine-sugar phosphotransferase activity\\
5e-04 & 104 & GO:0009401 & phosphoenolpyruvate-dependent sugar phosphotransferase system\\
\hline 
\end{tabular}
\par\end{centering}
\label{tab:go-events-recombining}
\end{table}
%\clearpage{}
%=============================================================================

%=============================================================================
% Table S. 
\begin{table}[!ht]
\caption{
{\bf Virulence genes inferred to have experienced recombining gene transfer
events.}}
\noindent \begin{centering}
\begin{tabular}{lrll}
\hline 
Direction & Fraction & Genes & Description\\ 
\hline 
SDE1-to-SDD & 0.97 & SpyM3\_1369 & putative serine threonine kinase\\
 & 0.25 & SpyM3\_1370 & putative phosphoprotein phosphatase\\
SDE2-to-SDD & 1 & SpyM3\_1522 & 3-oxoacyl-acyl carrier protein synthase II\\
 & 1 & SpyM3\_1523 & 3-ketoacyl-acyl-carrier-protein reductase\\
 & 1 & SpyM3\_1524 & acyl-carrier-protein S-malonyltransferase\\
SDD-to-SDE1 & 0.16 & SpyM3\_0499 & F0F1 ATP synthase subunit beta\\
SDD-to-SDE2 & 0.01 & SpyM3\_1767 & putative endopeptidase Clp ATP-binding chain C\\
SDD-to-SDE & 0.49 & SpyM3\_1180 & putative glucose kinase\\
SPY1-to-SDE1 & 0.13 & SpyM3\_0440 & putative calcium transporter\\
SPY1-to-SDE2 & 0.28 & SpyM3\_1755 & putative pyruvate formate-lyase activating enzyme\\
SPY2-to-SDE2 & 0.21 & SpyM3\_0420 & lysyl-tRNA synthetase\\
SPY2-to-SDE & 0.57 & SpyM3\_1806 & DNA mismatch repair protein MutS\\
SDE-to-SDD & 1 & SpyM3\_1188 & phosphopantetheine adenylyltransferase\\
 & 0.31 & SpyM3\_0036 & bifunctional acetaldehyde-CoA alcohol dehydrogenase\\
 & 0.26 & SpyM3\_0499 & F0F1 ATP synthase subunit beta\\
 & 0.05 & SpyM3\_0497 & F0F1 ATP synthase subunit alpha\\
SDE-to-SPY1 & 0.02 & SpyM3\_0440 & putative calcium transporter\\
SDE-to-SPY2 & 0.47 & SpyM3\_1366 & putative two-component response regulator\\
 & 0.19 & SpyM3\_1475 & N-acetylglucosamine-6-phosphate deacetylase\\
 & 0.09 & SpyM3\_1201 & putative two-component sensor histidine kinase\\
 & 0.02 & SpyM3\_1749 & putative pyruvate formate-lyase\\
SDE-to-SPY & 0.58 & SpyM3\_0499 & F0F1 ATP synthase subunit beta\\
 & 0.22 & SpyM3\_0497 & F0F1 ATP synthase subunit alpha\\
 & 0.08 & SpyM3\_1370 & putative phosphoprotein phosphatase\\
SPY-to-SDE1 & 0.86 & SpyM3\_0511 & ABC transporter ATP-binding protein\\
 & 0.18 & SpyM3\_1765 & chaperonin GroEL\\
SPY-to-SPY2 & 0.03 & SpyM3\_1749 & putative pyruvate formate-lyase\\
SPY-to-SDE & 0.98 & SpyM3\_0465 & putative dipeptidase\\
 & 0.84 & SpyM3\_1755 & putative pyruvate formate-lyase activating enzyme\\
 & 0.54 & SpyM3\_1749 & putative pyruvate formate-lyase\\
 & 0.3 & SpyM3\_0764 & putative spermidineputrescine ABC transporter ATP-binding protein\\
 & 0.28 & SpyM3\_0420 & lysyl-tRNA synthetase\\
 & 0.23 & SpyM3\_1182 & putative peroxide resistance protein\\
 & 0.22 & SpyM3\_0440 & putative calcium transporter\\
 & 0.07 & SpyM3\_1366 & putative two-component response regulator\\
 & 0.06 & SpyM3\_1180 & putative glucose kinase\\
\hline 
\end{tabular}
\par\end{centering}
\label{tab:virrecomb}
\end{table}
%\clearpage{}
%=============================================================================

%=============================================================================
% Table S. 
\begin{table}[!ht]
\caption{
{\bf Virulence genes inferred to have experienced horizontal gene transfer
events.}}
\noindent \begin{centering}
\begin{tabular}{ll}
\hline 
Genes & Gene Accessions\tabularnewline
\hline 
Streptococcal C5a peptidase & SDEG\_0933\\ 
Putative transposase & SpyM3\_1016, SpyM3\_1053, SpyM3\_1054\\
putative response regulator & SpyM3\_0174\\
putative maltose operon transcriptional repressor & SpyM3\_0982\\ 
putative histidine kinase & SpyM3\_0171\\
putative cell wall hydrolase & SpyM3\_0731, SpyM3\_0922, SpyM3\_1208\\
genome-encoded streptodornase & SDEG\_0541\\
prophage-associated deoxyribonuclease & SDEG\_1103\\
Protein F2 like fibronectin-binding protein (adhesin and invasin) & SpyM3\_0104\\
\hline 
\end{tabular}
\par\end{centering}
\label{tab:virhgt}
\end{table}
%\clearpage{}
%=============================================================================


%=============================================================================
% Table S6. Genes with evident recombining gene transfer.
\begin{table}[!ht]
\caption{
{\bf Genes and their parts transferred via homologous
recombining gene transfer.}}
{\footnotesize
\begin{tabular}{lllllll}
\hline 
Locus  & N$^a$  & S$^b$  & D$^c$ & F$^d$(\%)  & C$^e$(\%)  & gene 
product\tabularnewline
\hline 
SpyM3\_0074  &  & 4  & 0  & 20  & 100  & putative penicillin-binding protein 1b 
\tabularnewline
SpyM3\_0134  &  & 6  & 0  & 43, 0 & 79  & leucyl-tRNA synthetase \tabularnewline
SpyM3\_0156  &  & 0  & 4  & 77  & 100  & glucose-6-phosphate isomerase 
\tabularnewline
SpyM3\_0186  &  & 5  & 4  & 83  & 94  & putative glucose kinase \tabularnewline
SpyM3\_0192  & 1 & 6  & 5  & 14, 6 & 100  & ribulose-phosphate 3-epimerase 
\tabularnewline
SpyM3\_0193  & 1 & 6  & 5  & 99  & 100  & hypothetical protein\tabularnewline
SpyM3\_0197  &  & 6  & 5  & 7  & 100  & putative surface exclusion 
protein\tabularnewline
SpyM3\_0204  &  & 6  & 5  & 45  & 100  & hypothetical protein \tabularnewline
SpyM3\_0215  &  & 6  & 5  & 5, 26 & 100  & oligopeptide permease \tabularnewline
SpyM3\_0238  &  & 5  & 3  & 18  & 100  & hypothetical protein \tabularnewline
SpyM3\_0292  &  & 6  & 5  & 27  & 100  & putative tetrapyrrole methylase family 
protein \tabularnewline
SpyM3\_0429  &  & 2  & 1  & 0,22  & 100  & putative oligoendopeptidase 
F\tabularnewline
SpyM3\_0465  & 2 & 6  & 5  & 95  & 99  & putative dipeptidase \tabularnewline
SpyM3\_0466  & 2 & 6  & 5  & 16  & 100  & putative adhesion protein 
\tabularnewline
SpyM3\_0506  &  & 5  & 3  & 52  & 100  & phenylalanyl-tRNA synthetase subunit 
alpha \tabularnewline
SpyM3\_0580  & 3 & 3  & 0  & 3  & 100  & PTS system, fructose-specific IIABC 
component\tabularnewline
SpyM3\_0581  & 3 & 3  & 0  & 18  & 100  & putative peptidoglycan hydrolase 
\tabularnewline
SpyM3\_0581  &  & 6  & 5  & 11  & 100  & putative peptidoglycan hydrolase 
\tabularnewline
SpyM3\_0745  & 4 & 6  & 1  & 14  & 100  & putative two-component sensor 
histidine kinase \tabularnewline
SpyM3\_0746  & 4 & 6  & 1  & 55  & 100  & putative two-component response 
regulator \tabularnewline
SpyM3\_0785  & 5 & 4  & 1  & 78  & 92  & inorganic polyphosphate/ATP-NAD 
kinase\tabularnewline
SpyM3\_0786  & 5 & 4  & 1  & 100  & 100  & putative ribosomal large subunit 
pseudouridine synthase \tabularnewline
SpyM3\_0787  & 5 & 4  & 1  & 12  & 100  & phosphotransacetylase \tabularnewline
SpyM3\_1078  & 6 & 4  & 1  & 21  & 100  & putative cation/potassium uptake 
protein \tabularnewline
SpyM3\_1079  & 6 & 4  & 1  & 20  & 100  & putative ATP-dependent RNA helicase 
\tabularnewline
SpyM3\_1093  &  & 4  & 0  & 30  & 100  & putative heavy 
metal/cadmium-transporting ATPase \tabularnewline
SpyM3\_1151  &  & 3  & 1  & 34  & 100  & putative DNA repair and genetic 
recombination protein \tabularnewline
SpyM3\_1153  & 7 & 4  & 0  & 57  & 100  & putative hemolysin\tabularnewline
SpyM3\_1154  & 7 & 4  & 0  & 4  & 100  & putative geranyltranstransferase 
\tabularnewline
SpyM3\_1201  &  & 6  & 5  & 21  & 100  & putative two-component sensor 
histidine kinase \tabularnewline
SpyM3\_1267  & 8 & 6  & 5  & 7  & 100  & bifunctional methionine sulfoxide 
reductase A/B protein \tabularnewline
SpyM3\_1268  & 8 & 6  & 5  & 8  & 100  & hypothetical protein \tabularnewline
SpyM3\_1365  &  & 3  & 0  & 28  & 100  & putative Cof family 
protein/peptidyl-prolyl cis-trans isomerase, cyclophilin
type\tabularnewline
SpyM3\_1380  & 9 & 6  & 1  & 98  & 100  & putative Acetyl-CoA:acetoacetyl-CoA 
transferase b subunit \tabularnewline
SpyM3\_1381  & 9 & 6  & 1  & 100  & 100  & putative oxidoreductase 
\tabularnewline
SpyM3\_1382  & 9 & 6  & 1  & 23  & 38  & 3-hydroxybutyrate dehydrogenase 
\tabularnewline
SpyM3\_1512  &  & 4  & 0  & 59  & 100  & PTS system, mannose-specific IIC 
component \tabularnewline
SpyM3\_1518  & 10 & 1  & 2  & 63  & 100  & acetyl-CoA carboxylase subunit 
beta\tabularnewline
SpyM3\_1519  & 10 & 1  & 2  & 100  & 100  & acetyl-CoA carboxylase biotin 
carboxylase subunit \tabularnewline
SpyM3\_1520  & 10 & 1  & 2  & 100  & 100  & (3R)-hydroxymyristoyl-ACP 
dehydratase \tabularnewline
SpyM3\_1521  & 10 & 1  & 2  & 100  & 100  & acetyl-CoA carboxylase biotin 
carboxyl carrier protein subunit\tabularnewline
SpyM3\_1522  & 10 & 1  & 2  & 100  & 100  & 3-oxoacyl-(acyl carrier protein) 
synthase II \tabularnewline
SpyM3\_1523  & 10 & 1  & 2  & 100  & 100  & 3-ketoacyl-(acyl-carrier-protein) 
reductase \tabularnewline
SpyM3\_1524  & 10 & 1  & 2  & 100  & 100  & acyl-carrier-protein 
S-malonyltransferase\tabularnewline
SpyM3\_1525  & 10 & 1  & 2  & 100  & 100  & putative trans-2-enoyl-ACP 
reductase II \tabularnewline
SpyM3\_1561  &  & 6  & 5  & 11  & 97  & hypothetical protein \tabularnewline
SpyM3\_1577  &  & 4  & 1  & 11  & 100  & excinuclease ABC subunit A 
\tabularnewline
SpyM3\_1579  &  & 4  & 1  & 1  & 100  & hypothetical protein \tabularnewline
SpyM3\_1591  &  & 6  & 5  & 5  & 100  & ribonuclease HIII \tabularnewline
SpyM3\_1753  &  & 6  & 5  & 98  & 100  & putative transcriptional regulator 
\tabularnewline
SpyM3\_1753  & 11 & 3  & 1  & 53  & 100  & putative transcriptional regulator 
\tabularnewline
SpyM3\_1754  & 11 & 3  & 1  & 54  & 100  & putative transcriptional regulator 
\tabularnewline
SpyM3\_1754  &  & 6  & 5  & 95  & 100  & putative transcriptional regulator 
\tabularnewline
SpyM3\_1778  &  & 6  & 0  & 34  & 100  & putative cationic amino acid 
transporter protein \tabularnewline
SpyM3\_1809  &  & 4  & 0  & 5  & 100  & arginyl-tRNA synthetase \tabularnewline
SpyM3\_1813  & 12 & 3  & 0  & 19  & 100  & hypothetical protein \tabularnewline
SpyM3\_1814  & 12 & 3  & 0  & 29  & 100  & aspartyl-tRNA 
synthetase\tabularnewline
\hline
\end{tabular}

\begin{flushleft}
$^a$ Numbers represent a group of loci that are consecutive in gene order. Empty
cells denote stand alone loci,

$^b$ Donor branches of homologous recombinant edges, 0 (SDE1), 1 (SDE2), 2
(SDD), 3 (SPY1), 4 (SPY2), 5 (SDE), 6 (SPY), 7(SD), and 8(ROOT),

$^c$ Recipient branches of recombinant edges,

$^d$ Proportion of sites that the recombinant edge covers out of the sites in
the core alignment block, and

$^e$ Percentage of the sites in the core alignment block out of total sites of
the locus.
\end{flushleft}
}
\label{tab:genes-transfer}
\end{table}
\clearpage{}
%=============================================================================


%=============================================================================
% Table S8. ClonalOrigin - All Topologies
\begin{table}[!ht]
\caption{
{\bf Functional categories associated with 
recombination intensity based on all of the recombinant edge types.}}
\noindent \begin{centering}
\begin{tabular}{cccl}
\hline 
P-Value & Count & GO term & Description\tabularnewline
\hline 
4.596e-06 & 26 & GO:0019843 & rRNA binding\tabularnewline
5.395e-06 & 40 & GO:0030529 & ribonucleoprotein complex\tabularnewline
7.935e-06 & 39 & GO:0003735 & structural constituent of ribosome\tabularnewline
1.546e-05 & 81 & GO:0006412 & translation\tabularnewline
3.518e-05 & 66 & GO:0005840 & ribosome\tabularnewline
4.288e-05 & 22 & GO:0006633 & fatty acid biosynthetic process\tabularnewline
9.098e-05 & 20 & GO:0008610 & lipid biosynthetic process\tabularnewline
0.0003186 & 33 & GO:0008982 & protein-N(PI)-phosphohistidine-sugar phosphotransferase activity\tabularnewline
0.000357 & 45 & GO:0008643 & carbohydrate transport\tabularnewline
0.0006318 & 38 & GO:0009401 & phosphoenolpyruvate-dependent sugar phosphotransferase system\tabularnewline
0.0009886 & 105 & GO:0003723 & RNA binding\tabularnewline
\hline 
\end{tabular}
\par\end{centering}
\label{tab:functional-all}
\end{table}
%\clearpage{}
%=============================================================================


%=============================================================================
% Table S. Simulation study for the three population parameters.
\begin{table}[!ht]
\caption{
{\bf Functional categories asscoiated with 
recombination intensity based on the topology-not-changing recombinant edge
types.}}
\noindent \begin{centering}
\begin{tabular}{cccl}
\hline 
P-Value & Count & GO term & Description\tabularnewline
\hline 
6.515e-05 & 39 & GO:0003735 & structural constituent of ribosome\tabularnewline
0.00020986 & 40 & GO:0030529 & ribonucleoprotein complex\tabularnewline
0.00045049 & 66 & GO:0005840 & ribosome\tabularnewline
0.00052925 & 26 & GO:0019843 & rRNA binding\tabularnewline
\hline 
\end{tabular}
\par\end{centering}
\label{tab:functional-notopology}
\end{table}
%\clearpage{}
%=============================================================================




%=============================================================================
% Table S. ClonalOrigin - SDE-to-SPY
%\begin{table}
%\caption{\label{tab:functional-sde-spy}Functional categories asscoiated with 
%recombination intensity based on SDE-to-SPY recombinant edge types.}
%\noindent \begin{centering}
%\begin{tabular}{cccl}
%\hline 
%P-Value & Count & GO term & Description\tabularnewline
%\hline 
%1.854e-05 & 38 & GO:0009401 & phosphoenolpyruvate-dependent sugar phosphotransferase system\tabularnewline
%2.703e-05 & 81 & GO:0006412 & translation\tabularnewline
%6.482e-05 & 33 & GO:0008982 & protein-N(PI)-phosphohistidine-sugar phosphotransferase activity\tabularnewline
%9.777e-05 & 105 & GO:0003723 & RNA binding\tabularnewline
%0.00012793 & 40 & GO:0030529 & ribonucleoprotein complex\tabularnewline
%0.00025176 & 26 & GO:0019843 & rRNA binding\tabularnewline
%0.00031236 & 72 & GO:0005975 & carbohydrate metabolic process\tabularnewline
%0.00044756 & 317 & GO:0005737 & cytoplasm\tabularnewline
%0.00059675 & 45 & GO:0008643 & carbohydrate transport\tabularnewline
%0.00066874 & 39 & GO:0003735 & structural constituent of ribosome\tabularnewline
%0.00135304 & 10 & GO:0006098 & pentose-phosphate shunt\tabularnewline
%\hline 
%\end{tabular}
%\par\end{centering}
%\end{table}
%\clearpage{}
%=============================================================================

%=============================================================================
% Table S. ClonalOrigin - SPY-to-SDE
%\begin{table}
%\caption{\label{tab:functional-spy-sde}Functional categories asscoiated with 
%recombination intensity based on SPY-to-SDE recombinant edge types.}
%\noindent \begin{centering}
%\begin{tabular}{cccl}
%\hline 
%P-Value & Count & GO term & Description\tabularnewline
%\hline 
%1.106e-06 & 45 & GO:0008643 & carbohydrate transport\tabularnewline
%1.511e-06 & 33 & GO:0008982 & protein-N(PI)-phosphohistidine-sugar phosphotransferase activity\tabularnewline
%1.595e-06 & 38 & GO:0009401 & phosphoenolpyruvate-dependent sugar phosphotransferase system\tabularnewline
%1.443e-05 & 187 & GO:0006810 & transport\tabularnewline
%0.0001106 & 371 & GO:0016740 & transferase activity\tabularnewline
%0.0003596 & 72 & GO:0005975 & carbohydrate metabolic process\tabularnewline
%0.0004023 & 13 & GO:0006814 & sodium ion transport\tabularnewline
%0.000823 & 317 & GO:0005737 & cytoplasm\tabularnewline
%0.001142 & 12 & GO:0031402 & sodium ion binding\tabularnewline
%\hline 
%\end{tabular}
%\par\end{centering}
%\end{table}
%\clearpage{}
%=============================================================================

%=============================================================================
% Table S. Mowgli - Duplication 
%\begin{table}
%\caption{\label{tab:go-events-duplication}Functional categories enriched in 
%gene duplication events that are inferred using the parsimony-based approach.}
%\noindent \begin{centering}
%\begin{tabular}{cccl}
%\hline 
%P-value & Count & GO-term & Description \\
%\hline 
%4e-81 &  67 & GO:0006313 & transposition, DNA-mediated\\
%5e-80 &  59 & GO:0004803 & transposase activity\\
%6e-70 &  23 & GO:0032196 & transposition\\
%2e-58 &  52 & GO:0015074 & DNA integration\\
%8e-25 &  86 & GO:0006310 & DNA recombination\\
%1e-20 &  36 & GO:0016987 & sigma factor activity\\
%1e-20 &  36 & GO:0006352 & transcription initiation\\
%2e-19 & 197 & GO:0003676 & nucleic acid binding\\
%2e-10 & 527 & GO:0003677 & DNA binding\\
%4e-06 & 142 & GO:0043565 & sequence-specific DNA binding\\
%\hline 
%\end{tabular}
%\par\end{centering}
%\end{table}
%\clearpage{}
%=============================================================================

%=============================================================================
% Table S. Mowgli - Loss
%\begin{table}
%\caption{\label{tab:go-events-loss}Functional categories enriched in 
%gene loss events that are inferred using the parsimony-based approach.}
%\noindent \begin{centering}
%\begin{tabular}{cccl}
%\hline 
%P-value & Count & GO-term & Description \\
%\hline 
%1e-06 &  67 & GO:0006313 & transposition, DNA-mediated\\
%2e-05 &  59 & GO:0004803 & transposase activity\\
%5e-05 &  52 & GO:0015074 & DNA integration\\
%5e-04 &  23 & GO:0032196 & transposition\\
%5e-04 &  10 & GO:0046873 & metal ion transmembrane transporter activity\\
%7e-04 &  36 & GO:0016987 & sigma factor activity\\
%7e-04 &  36 & GO:0006352 & transcription initiation\\
%1e-03 &  31 & GO:0006814 & sodium ion transport\\
%3e-03 &  86 & GO:0006310 & DNA recombination\\
%6e-03 & 197 & GO:0003676 & nucleic acid binding\\
%\hline 
%\end{tabular}
%\par\end{centering}
%\end{table}
%\clearpage{}
%=============================================================================


%=============================================================================
% Table S15. Simulation study for the three population parameters.
\begin{table}[!ht]
\caption{
{\bf Simulation study of the three population parameters.}}
\noindent \begin{centering}
\begin{tabular}{ccc}
\hline
 & True & Estimates (Standard Deviation)\tabularnewline
\hline
Mutation rate per site & 0.081 & 0.081 (0.0011)\tabularnewline
Recombination rate per site & 0.012 & 0.011 (0.0003)\tabularnewline
Recombinant tract length & 744 bp & 919 bp (28)\tabularnewline
\hline
\end{tabular}
\par\end{centering}
\label{tab:sim-three}
\end{table}
%=============================================================================


\clearpage{}
\section*{Supplementary Figures}

%=============================================================================
% Figure S1: Mauve alignment of the five genomes.
\begin{figure}[!ht]
\begin{center}
\includegraphics[scale=0.25]{figures/mauve}
\end{center}
\caption{\label{fig:mauve}
{\bf The genome alignment of the five genomes.} The top
first and second rows are SDE1 and SDE2 genomes, respectively. The third is SDD
genome. The fourth and fifth rows are SPY1 and SPY2 genomes, respectively.}
\end{figure}
\clearpage{}
%=============================================================================

%=============================================================================
% Figure S: Histogram of block sizes
%\begin{figure}
%\begin{center}
%\includegraphics[width=5in]{figures/blocksize}
%\end{center}
%\caption{\label{fig:blocksize}Length distribution of the 274 core alignment blocks.}
%\end{figure}
%\clearpage{}%
%=============================================================================

%=============================================================================
% Figure S2. Three population parameters.
\begin{figure}[!ht]
\begin{center}
\subfloat[]{\includegraphics[scale=0.5]{figures/scatter-plot-parameter-1\lyxdot out\lyxdot theta}}

\subfloat[]{\includegraphics[scale=0.5]{figures/scatter-plot-parameter-1\lyxdot out\lyxdot rho}}

\subfloat[]{\includegraphics[scale=0.5]{figures/scatter-plot-parameter-1\lyxdot out\lyxdot delta}}
\end{center}
\caption{
{\bf Three scatter plots of mutation rate, recombination
rate, and average tract length for the blocks.} The plus signs are mean values
for each block. The density of each value for each block is depicted by black
clouds.  The red dashed lines are the global block-length weighted median of the
three parameter estimates.}
\end{figure}
\label{fig:scatter3}
\clearpage{}%
%=============================================================================

%=============================================================================
% Figure S3: gene family sizes
% FIGURE - gene family sizes
\begin{figure}[!ht]
\begin{center}
\includegraphics[width=5in]{figures/famsizes}
\end{center}
\caption{{\bf Distribution of gene family sizes.}}
\label{fig:famsizes}
\end{figure}
\clearpage{}
%=============================================================================

%=============================================================================
% FIGURE S4. - HGT Transfer heatmap
\begin{figure}[!ht]
\begin{center}
\includegraphics[scale=0.45]{figures/heatmap-hgtedge}
\end{center}
\caption{
{\bf Heat map of horizontal gene transfers using the six genomes.}
Each cell of the color matrix shows the number of inferred non-recombining
horizontal gene transfer events between the different branches of the species
tree. Black colored cells represent no inferred events.}
\label{fig:hgt-heatmap}
\end{figure}
\clearpage{}%
%=============================================================================

%=============================================================================
% Figure S5: recombination heatmap from mowgli
\begin{figure}[!ht]
\begin{center}
\includegraphics[scale=0.45]{figures/heatmap-recedge-mowgli}
\end{center}
\caption{
{\bf Heat map of recombining gene transfers using the six genomes.}
Each cell of the color matrix shows the number of inferred recombining
gene transfer events between the different branches of the species
tree. Black colored cells represent no inferred events.}
\label{fig:mowgli-recomb-heatmap}
\end{figure}
\clearpage{}
%=============================================================================

%=============================================================================
% Figure S6: Comparison of ClonalOrigin and Mowgli analysis
\begin{figure}[!ht]
\begin{center}
\includegraphics[width=5in]{figures/compareMowgliCoRecombining}
\end{center}
\caption{
{\bf Plot of recombination intensity against number
of recombining gene transfer.}}
\label{fig:cmpcomowgli}
\end{figure}
\clearpage{}
%=============================================================================

%=============================================================================
% Figure S7: HGT Comparison of ClonalOrigin and Mowgli analysis
\begin{figure}[!ht]
\begin{center}
\includegraphics[width=5in]{figures/compareMowgliCoHgt}
\end{center}
\caption{
{\bf Plot of recombination intensity against number
of horizontal gene transfer.}}
\label{fig:cmpcomowglihgt}
\end{figure}
\clearpage{}
%=============================================================================

%=============================================================================
% Figure S5: Scatter plot of rho and log(delta)
% \begin{figure}
% \begin{center}
% \includegraphics[width=5in]{figures/rho-vs-delta}
% \end{center}
% \caption{Plot of recombination rate and logarithm of average tract length
% of the posterior samples.}
% \label{fig:rhologdelta}
% \end{figure}
% \clearpage{}%
%=============================================================================

%=============================================================================
% Figure S8: View of UCSC genome browser for bacterial gene transfer.
\begin{figure}[!ht]
\includegraphics[scale=0.8]{figures/ucsc}
\caption{
{\bf UCSC genome browser tracks for gene transfer study of
SPY and SDE.}}
\label{fig:ucsc}
\end{figure}
\clearpage{}%
%=============================================================================

%=============================================================================
% Figure S6. A Recombinant Tree of ClonalOrigin Model
\begin{figure}[!ht]
\includegraphics[scale=0.42]{figures/clonalorigin}
\caption{
{\bf Recombinant tree in the \texttt{ClonalOrigin} model.}
(A) A recombinant tree relating the five genomes with four recombinant edges.
(B) A genomic region of size 100 base pairs with subregions affected by the
four recombinant edges. Genomic positions are evenly marked by ticks every 10-th
base pair.  For each edge type and each site, we recorded the sampling frequency
of a recombinant edge of the type at that site.  The posterior probability that
a site experiences influx from a donor branch using a single recombinant tree.
Summing these posteriors over a certain set of edge types, we obtain a measure
of recombination intensity along the genome.  Each segment from b to g are
scored by the number of recombinant edges per site.  (C-G) Local gene trees
induced by the recombinant tree are illustrated for each region covered by
combinations of recombinant edges along the genomic region. For example, the
local gene tree (E) is created by three recombinant edges in the subregion
between 51 base pairs and 70 base pairs. The local gene tree (C) is the same as
the species tree (A) in their topology.  Because there are 105 possible rooted
labeled topologies relating five taxa, we could count topologies.}
\label{fig:clonalorigin}
\end{figure}
\clearpage{}%
%=============================================================================

%=============================================================================
% Figure S13:
\begin{figure}[!ht]
\includegraphics{figures/ri1}
\caption{
{\bf The surrogate measure of recombination intensity
of genes.} Each point represents a pair of a true value of average number of
recombinant edges and its estimated value for a gene. The dashed line is with
slope of 1 and intercept of 0.}
\label{fig:ri1}
\end{figure}

\clearpage{}%
%=============================================================================


%=============================================================================
% Figure S13: 
%\begin{figure}
%\subfloat[]{\includegraphics[scale=0.5]{figures/h1-R}}
%\subfloat[]{\includegraphics[scale=0.5]{figures/h1-R-inset}}
%\caption{\label{fig:h1}Numer of recombination events. The diagonal dashed
%lines are with slope of 1 and intercept of 0. For each order
%pairs of branches of a reference species tree the true number of recombination
%events is computed as the mean of the 10 replicates of recombinant
%trees from the simulation. With ten estimates of each replicate total
%number of 100 estimats are shown as intervals of 90\% of quantile
%range, or the lower value of the interval is the 5\% quantile, and
%the upper one is the 95\% quantile. }
%\end{figure}
%\clearpage{}%

%=============================================================================
%=============================================================================
% Figure S:
%\begin{figure}
%\subfloat[]{\includegraphics[scale=0.5]{figures/h2-R}}
%\subfloat[]{\includegraphics[scale=0.5]{figures/h2-R-inset}}

%\subfloat[]{\includegraphics[scale=0.5]{figures/h2-R-inset2}}

%\caption{\label{fig:h2}Comparison of number of recombinations between real
%and simulated data sets. The diagonal dashed lines are with
%slope of 1 and intercept of 0. For each order pairs of branches of
%a reference species tree the number of recombination events is estimated
%from the real data set. Total number of 100 estimats from a simulated
%data set are shown as intervals of 90\% of quantile range, or the
%lower value of the interval is the 5\% quantile, and the upper one
%is the 95\% quantile. }
%\end{figure}
%\clearpage{}%
%=============================================================================


\clearpage{}
\section*{Supplementary Information}

\subsection*{Summarization of posterior samples for local gene trees and recombination intensity}

Because the procedure of summarizing posterior samples of \textit{recombinant
tree}, which will be defined in a moment, was not detailed in the user guide at
the time when we used \texttt{ClonalOrigin}, we elected to use Figure
S\ref{fig:clonalorigin} to illustrate a recombinant tree in order to clarify
summarization of \textit{recombinant history} posterior samples.  A
recombination history is represented by a \textit{recombinant tree}, which is a
species tree augmented with \textit{recombinant edges}.  A recombinant edge
connect two species tree branches.  Each recombinant edge is defined by the time
and branch of origin and destination, and the genomic interval that it affects.
The branch of origin is referred to as the \textit{donor} and the destination
branch is referred to as the \textit{recipient}. The type of a recombinant edge
is determined by the donor and recipient species tree branches.  A recombinant
tree can be viewed as a restricted version of an ancestral recombination graph
(ARG), and as such it induces a local genealogy for each site in the genomic
block it describes.  Didelot et al.\ \citet{Didelot2010} presented two procedures of summarizing
recombinant trees by viewing trees from two different angles: by counting types
of recombinant edges regardless of genomic positions affected by the edges, and
by measuring recombinant edge changes to the species tree in site-wise manner.
By similarly following their approaches we quantified additional aspects of
recombinant trees; we counted gene tree topologies for each site, and also
quantified recombination intensity along the genome.

The recombination intensity allowed us to find genomic regions enriched for
recombination events in general, or regions enriched for recombination events of
certain types (e.g., between SPY and SDE).  We noted that the recombination
intensity was not normalized to the interval of $[0,1]$. Rather than giving
probabilistic meanings to the recombination intensity we interpreted it as an
average number of recombinant edge types per site.  The measure appeared to act
as an effective indicator of recombination intensity (see subsection of
\textbf{Simulation Study}).  Our simulation study indicated that edges of
certain types were better indicators of recombination events than others.  For
instance, recombinant edges which resulted in a recombinant tree with the same
topology as the species tree appeared to be more prone to spurious sampling.
Such edges were ones where the donor was the same branch, sister branch, or
parent branch of the recipient branch.  Therefore, in some cases we elected to
remove such edges from consideration when measuring recombination intensity.  
% The tree topologies would show more effectively regions where
% recombination occurs along the genome (see subsection of \textbf{UCSC Genome
% Browser}).

\subsection*{Simulation 1 - three population parameters}

We describe a procedure of simulations for checking if the three
population parameters were accurately recovered using the model-based approach.
Ten recombinant trees were sampled given the three parameters (Table
S\ref{tab:three}) and the species tree (Figure S\ref{fig:tree5}) set to the
values inferred from the five bacterial genomes.  Each of the recombinant trees
was used to simulate the Jukes-Cantor model for DNA sequence evolution to create
274 blocks of alignments.  The block lengths configuration was the same as that
of the blocks used in the analysis of the five genomes.  

% s15
% FIXME: Newick tree in text in a file.
% FIXME: Publish the result and the data set with the code.

\subsection*{Simulation 2 - numbers of recombinant edges from prior}

We first sampled ten recombinant trees given the mutation rate, recombination
rate, and tract length (Table S\ref{tab:three}) and the species tree (Figure
S\ref{fig:tree5}), and then created ten independent simulated data sets with
the Jukes-Cantor model for DNA sequence evolution using each of the recombinant
trees. This consisted of 100 replicates of data sets.  The recombinant trees
were created from the prior distribution and recombinant edges were added to the
species tree with no particular directions of recombining gene transfers between
species tree branches.  We computed the fraction of the mean number of recombinant
edges for a pair of donor and recipient species tree branches over the simulated
10 recombinant trees against the median of recombinant edges estimated from the
10 simulated data sets, and then took the logarithm of the fraction to base 2. We
also computed 5\% and 95\% quantiles to find the 90\% quantile range for the log
base 2 fraction.  We expect that the log base 2 fraction would near zero. We also
wished to see if how the recombinant edge count estimates from the real data set
deviate from the prior. Instead of using the number of recombinant edges from
the ten recombinant trees, we used the numbers of recombinant edges that were
estimated from the real data. 

We tested if our measure of recombination intensity provided a reliable estimate
of the true intensity. We artificially annotated our synthetic blocks to genes
according to the real annotations from SPY1 genome.  For each gene in each of
the simulated 100 data sets, we plotted the average estimated recombination
intensity along that gene against the average number of different recombinant
edge types from the ten simulated recombinant tree. 

% s16 


% s16
% FIXME 1: IG: RESULTS FROM THIS POINT ON REQUIRE SOME REFINEMENT. SC: Let's see.
% FIXME 2: ``TRUE" TO BE EXPECTED VALUE
% RATHER THAN AVERAGE OVER THE 10 REC.  TREES. SC: OKAY. LET'S FIND THE PRIOR
% EXPECTED NUMBER OF EVENTS. THIS MUST BE GIVEN SOMEWHERE}.  
% FIXME 3: CONSIDER SHOWING DETAILED COMPARISON OF ESTIMATED VALUES TO ONES 
% EXPECTED FROM THE ACTUAL REC. TREES. THESE DON'T LOOK AS NICE, BUT THEY MIGHT 
% BE ABLE TO DEMONSTRATE OUR ABILITY TO DETECT DEPARTURE FROM EXPECTATION. SC:
% THIS MAY BE ABOUT CHANGING AVERAGE TO PRIOR EXPECTED NUMBER OF EVENTS.
% FIXME 4: WE SHOULD PROBABLY CONSIDER ALTERNATIVE WAYS OF USING THESE RESULTS TO 
% ARGUE THAT OUR DIRECTIONAL SIGNAL IS REAL. SC: I MIGHT CONSIDER SIMULATIONS AND
% ARTIFICIALLY DOUBLE OR TRIPLE THE RECOMBINATION EVENTS OF ONE OF TYPES.

\subsection*{Simulation 3 - number of recombinant edges from posterior}

We used 100 recombinant trees sampled from the posterior distribution
of the recombiant tree instead of using simulated recombinant trees.
We created one simulated data set for each of the 100 recombinant trees with
the Jukes-Cantor model for DNA sequence evolution 
using the mutation rate estimate (Table S\ref{tab:three}).
This consisted of 100 replicates of data sets.  
We follow the same computation of the fraction as in the simulation 2 with 
the numbers of recombinant edges estimated from the real data
as numerators. 

% s17

\end{document}




\subsection*{Recombination drived more substitutions than did point mutation}

Five genomes except \textit{S.\ equi} ssp.\textit{\ equi} strain 4047 were
chosen (Table S\ref{tab:genome}) to study recombining gene transfer using the
model-based approach.  A genome alignment of the five genomes showed two
distinct genome structures of SPY and SDE (Figure S\ref{fig:mauve}). The SDE and
SDD genomes were larger in length than those of SPY. Although the graphical view
of the genome alignment did not clearly show whether SDD was different from
either of SDE1 and SDE2 in the genome organization, a spcies tree inferred from
the genome alignment showed that the external SDD branch was attached to the
internal clade branch of SDE1 and SDE2 (Figure S\ref{fig:tree5}). The
distribution of the alignment block lengths was skewed to the right with mean of
4188 base pairs long (Figure S\ref{fig:blocksize}), which supported that the
model assumption of geometric block length distribution might be plausible in
\texttt{ClonalOrigin} method (see section \textbf{Materials and Methods} and
\citet{Didelot2010}).  The relative strength ($r/m$) of recombination compared
to mutations indicated that recombination-driven substitutions affected more
sites of the core genomes than did mutation-driven substitutions (Table
S\ref{tab:clonalframe}). This could imply that in the \textit{Streptococcus}
species bacterial recombination process appeared to accelerate molecular
evolution of more sites than did point mutations process.

\subsection*{Recombination hotspots were observed in genomes}

Population mutation rate did not appear to change much across the genome of SDE1 
(Figure S\ref{fig:scatter3}).  We found a few spots of genomic regions where the
mutation rates were larger than twice the global estimate of mutation rate.  The
recombinant tract length changed in a way that we could not note any particular
positions with notable tract length values compared with other positions along
the genome.  As recombinant tract lengths were smaller, recombination rates per
site seemed larger (Figure S\ref{fig:rhologdelta}).  The recombination rate
appeared to change more dynamically. Especially, a genomic region of SDE1 near
200 kilobase pairs from the origin of replication contained alignment blocks with
bell-shaped recombination rates.  

\subsection*{SPY and SDE were more closely related in about tenth of the core genome}

There are 105 possible rooted labeled topologies relating five taxa.  We expected the
dominant topology across sampled genealogies to be concordant with that of the
species tree.  Indeed, two thirds of the local gene tree topologies (see section
\textbf{Materials and Methods}) along the alignment blocks were the same as the
species tree, which allowed us to be more confident of the species tree inferred
using the alignment blocks (Table S\ref{tab:Gene-tree-topologies}).  Other
frequently sampled topologies were expected to be potentially informative about
common recombination events.  The second most frequent topology (9.3\%) was one
where SDD was the outgroup, and SPY and SDE were sister clades. Therefore, SPY
and SDE appeared to be more closely related with each other than each to SDD's
genome in about tenth of the core genome. The other topologies had less than 4\%
of proportions.  

\subsection*{Unbalaned recombining gene flow from SPY to SDE in genome-wide scale}

The heat map of Figure \ref{fig:Heatmap-of-recombination} (see Tables
S\ref{tab:heatmap} and S\ref{tab:obsheatmap} for the heat map values and the
observed number of recombinant edges) shows the logarithm base 2 fractions of the
\textit{a posteriori} number of recombinant edges for a pair of donor and
recipient species tree branches and their corresponding expected values given
the species tree (Figure S\ref{fig:tree5}) and the three population parameters
(Table S\ref{tab:three}).  Because recombinant edges from any species tree
branch to the root branch are impossible (see Figure S\ref{fig:clonalorigin}),
the column for the root as recipient was colored black, indicating invalid
recombinant edges in Figure \ref{fig:Heatmap-of-recombination}.  
Similarly, all of the cells for which a recipient branch was
older than the donor, and the donor and recipient branches did not share any contemporary time were
colored black. The diagonal cells were colored gray, indicating no significant
number of recombinant edges, because the observed and expected numbers of
recombinant edges were similar. As recombinant edges that start from and end in
the same species tree branch would not change the species tree topology, which
might not
change the likelihood much, the cells would be colored gray.  Because SDD is a
strictly veterinary pathogenic bacteria, we did not expect that SDD received
much gene transfer from either SPY1 or SPY2: the logarithm base 2 fraction from SPY1 to
SDD was -2.20, and that from SPY2 to SDD was -1.97.  On the contrary, the
recombing gene transfer from SDE1 to SDD was 0.756, and that from SDE2 to SDD
was 0.698. This relatively higher recombining gene transfer from two SDE
individuals to SDD compared with that from two SPY individuals to SDD reflected
the fact
that the SDD was more closely related to SDE than to SPY in most of the genomes,
and the genomic organization between SDD and SDE1/SDE2 were much similar
(Figure S\ref{fig:mauve}).  However, the gene transfer from SDD to either SDE1
or SDE2 was less frequent than the case of reverse direction; SDD effectively
served as a recipient rather than a donor. It was unclear why SDD received
more genes from SDE1 or SDE2 than did SDE1 or SDE2 from SDD.  

Because it was expected that the recombining gene transfer within a clade would
be more frequent than that between clades, not surprisingly the recombining gene
transfer logarithm base 2 fractions between SDE1 and SDE2 (1.17 and 1.20) were about
twice as large as those from SDE1 and SDE2 to SPY1 and SPY2 (0.67, 0.66, 0.60,
and 0.39).  The intensities of recombining gene transfer from external branches
of SDE1 and SDE2 to external branches of SPY1 and SPY2 was similar to those from
SDE1 or SDE2 to SDD (0.76 and 0.70).  Therefore, either SPY1 or SPY2, and SDD as
recipient appeared to be no different from the view point of SDE1 or SDE2 as
donor since the similar values of recombining gene transfer intensity from SDE1
or SDE2 to one of the other three external branches. Surprisingly, the recombining
gene transfer logarithm base 2 fractions between SPY1 and SPY2 (1.06 and 1.07) was
comparable with, yet even less than those from SPY1 and SPY2 to SDE1 and SDE2
(1.25, 1.25, 1.17, and 1.29).  The recombining gene transfer from a clade SPY1 
or SPY2 to a different clade SDE1 or SDE2 seemed to be effectively no different
from that within the clade of SPY1 and SPY2.  The two external branches of SDE
received genes from either SPY1 or SPY2 via recombining gene transfer as if SDE1
and SDE2 were in the same clade as SPY1 and SPY2.  This was more surprising
because the bacterial recombination between SPY and SDE seemed to happen more
often than that between SDE and SDD even though the genome configurations of SDE
and SDD were more similar than each to SPY.  However, this result could be
concordant with the fact that SPY and SDE shared the same ecological niche
despite their dissimilar genomic organization.  The gene transfer logarithm base 2
fraction from SPY1 and SPY2 to SDE branch was also larger than any type of
recombining gene transfer: 1.88 and 1.94, respectively.  Additionally we found
58 genes for which a recombinant edge of certain types were sampled with
probability at least 0.9 by focusing on specific types of recombinant edges
(Table S\ref{tab:genes-transfer}).  Many of the genes in the list were
transferred from SPY to SDE rather than the reverse direction overall.  All
things considered, this epitomized that the recombining gene transfer was biased
toward the net direction from SPY to SDE.  

We tried in vain to quantify differential horizontal gene transfer between SPY
and SDE with the parsimony-based approach (Figure \ref{fig:hgt-heatmap}).  Among
the clade-crossing transfers, we did not see any significant enrichment for a
specific direction, as opposed to the result for recombining transfers from the
model-based analysis.  SDE1 and SDE2 shared the most events (26 and 34) compared
to other sample pairs.  SDD appeared to acquire many genes from the ancestral
SDE branch.  Transfer events between the SPY and SDE clades were inferred in
smaller numbers ($<6$) compared to those within each clade. We were puzzled by
this finding because virulence genes were found to be associated with horizontal
gene transfer, and yet we could not show net directionality of horizontal gene
transfer between SPY and SDE.  However, we inferred relatively more recombining
gene transfer events from the SPY branch to SDE branch even in the
parsimony-based approach (Figure S\ref{fig:mowgli-recomb-heatmap}).

\subsection*{Reduction of genes and divergence of SPY and SDE species}

The parsimony-based approach estimated 304 duplications, 944 recombining gene
transfers, and 294 horizontal gene transfers, comprising 23.7\% of total
transfer events (Figure \ref{fig:Gene-duplication-loss}).  Among the 2314 gene
families the largest family consisted of 59 genes, while nearly half of gene
families (1066) had exactly one gene per species, or six genes total (Figure
S\ref{fig:famsizes}).  Numbers of genes at the five ancestors were inferred to
be smaller than those at the extant species.  SDE lineages from the root
appeared to have increased in the number of genes while the common ancestor of
SPY1 and SPY2 decreased in the number of genes.

The four external lineages of SDE1, SDE2, SDD, and SPY1 had almost equal numbers
of gene gain and loss events while the four lineages of SP, SD, SPY2, and SEE
had larger numbers of gene gain events than those of gene loss events. In the
two lineages of SP and SEE we might have inferred gene gain and loss events less
confidently than events in other lineages because of lack of outgroup lineages
to compare with the two lineages. The number of gene gain events in SPY2 lineage
seemed to be large compared with those of gene gain events in the other
lineages. On the contrary, two internal branches of SPY and SDE had smaller
numbers of gene gain events than those of gene loss events.  This might indicate
that SPY and SDE lineages had reduced number of genes to specialize in their
specific niches after each split from their corresponding direct ancestors.

Duplications in extant lineages were more frequent than those in the other
lineages.  Although the number of duplication events along SPY2 branch seemed
smaller than those of the other extant lineages, duplication events appeared to
be most often species-specific.  Within clades we found that duplications made
up only a small fraction of the events  (10.2\%), while horizontal transfers and
losses were significantly more frequent (41.7\% and 48.0\%, respectively).  The
extant species of SDD and two SDE's were the most expanded, both in terms of
duplications and in-bound transfers. This was apparently related with the larger
genome sizes of the two species than that of SPY.  
The numbers of recombining gene transfer events tended to be larger than those
of horizontal gene transfer events except SD lineage.  Numbers of recombining
gene transfer in the extant lineages might have been larger than those in the
internal branches because the number of recombining gene transfer events must
had been affected by the availability of more closely related individual
genomes.  

\subsection*{Virulence genes via recombining gene transfer}

About 10\% of the genes comprised of virulence facters, about half of which were
found in the core genome of the five genome alignment \citep{Suzuki2011}.  Using
customized UCSC genome browser tracks available at
\url{http://strep-genome.bscb.cornell.edu} we located virulence genes with
notable posterior recombination probability.  These genes were by no means a
comprehensive list of virulence factors that might have been transferred via
homologous recombination.  We wished to showcase virulence genes using the
result of our study of gene transfer.

Stressful host conditions for pathogenic bacteria can include heat shock,
starvation, and iron depletion, etc.  Iron uptake of bacteria was known to be
essential for survival of bacteria under the hostile host condition. We found
that parts of putative calcium transporter (MgtA) in SDE1 could have been
transferred from SPY1 (SpyM3\_0440).  An extracellular protein, putative
dipeptidase (SpyM3\_0465) that could be important for bacteria to scavenge
dipeptides from their host, was inferred to be transferred from SPY to SDE: the
whole gene was transferred, which was a case for the finding that virulence
genes tended to be transferred as a whole unit \citep{Chan2009}.  Another
example of whole gene transfer was superoxide dismutase that was reported to be
associated with virulence of \textit{Streptococcus agalactiae}
\citep{Poyart2001}: the whole gene including its 5' UTR, (SpyM3\_1071), was
affected by recombining gene transfer from SPY to SDE.  Some of the genes were
only affected in their parts.  The middle parts of an adhesins gene, putative
internalin A precursor (SpyM3\_1035), was found to be affected by recombining
gene transfer from SPY to SDE.  The middle part of a chaperonin GroEL, putative
heat shock protein (SpyM3\_1765), was inferred to transfer from SPY to SDE1:
increase of GroEL expression was known to allow for the bacteria to survive the
hostile host environment of acid shock.  Parts of the gene (SpyM3\_1749),
pyruvate formate lyase that was recently reported to have virulence
\citep{Yesilkaya2009}, seemed to be affected by recombining gene transfer.
Under starvation condition bacteria must use more effective energy metabolism: a
putative glucose kinase, (SpyM3\_1180), were inferred to be transferred from SPY
to SDE to adapt SDE's energy metabolism to survive such harsh condition.  We
also found that an MLST locus, called DNA mismatch repair protein (MutS),
appeared to be affected by recombining gene transfer. 

\subsection*{Association of virulence genes and horizontal gene transfer}

Horizontal gene transfers were found to be significantly associated with
virulence genes (P-value of $<1.33 \times 10^{-11}$) whereas recombining gene
transfers were not enriched for virulence genes (P-value of $>0.90$) using the
parsimony-based approach.  This was also consistent with the fact that we could
not find the association of recombining gene transfer with virulence genes in
the model-based approach.  This supported more evidence that non-recombining
gene transfer served as a more effective means by which SDE or SPY acquired
virulence genes, compared to recombining gene transfer.  Families with high
numbers of duplications were also significantly enriched for virulence genes
(P-value of $<7.13 \times 10^{-3}$) whereas families with high losses showed no
enrichment (P-value of $>0.54$). 

Roughly, half of the virulence genes were in the pan genome as mentioned above.
We did not attempt to comprehensively list virulence genes that were inferred to
be transferred horizontally or non-recombining fashion. We rather sort virulence
genes by the number of horizontal gene transfer events to list the first several
genes with high ranks.  Streptococcal C5a peptidase (\textit{scpB}, SDEG\_0933)
was inferred to have six horizontal gene transfer.  Putative transposase were
found to have a relatively large number of horizontal gene transfer events
(SpyM3\_1016, SpyM3\_1053, SpyM3\_1054). Other genes with relatively large
number of horizontal gene transfer events included putative response regulator
(SpyM3\_0174), putative maltose operon transcriptional repressor (SpyM3\_0982),
putative histidine kinase (SpyM3\_0171), and putative cell wall hydrolase
(SpyM3\_0731, SpyM3\_0922, SpyM3\_1208).  Virulence genes encoding putative
nucleases with a secretion signal peptide were also found to have had
experienced horizontal gene transfer: genome-encoded streptodornase
(SDEG\_0541), and prophage-associated deoxyribonuclease (SDEG\_1103).  Protein
F2 like fibronectin-binding protein (SpyM3\_0104), an adhesin and invasin, was
inferred to horizontally transferred.

\subsection*{Model-based approach or parsimony-based one for recombining gene
transfer study}

A full model-based approach would be desirable in analyzing data for testing
relevant hypotheses more systematically. Model complexity might not allow easy
access to the implementation of a model. Our study involved a model-based
approach and a parsimony-based one for understanding evolution in the core
genomes of several bacteria individuals. Because of inability of the model-based
method for studying the pan genomes we had to resort to the parsimony-based
method. However, the parsimony-based method was also applied to the core
genomes. We wished to compare the two methods in their results of the core
genome analyses.  Because of two independent analyses on the same data set, this
would allow us to be more confident of the results if the two analyses performed
similarly. Generally, phylogenetic methods of detecting gene transfers often take
advantage of discrepancy between a reference species tree and gene trees. The
two approaches that we employed in this study were no exception. We found a few
similar results in both methods; firstly, functional categories that were
associated with recombining gene transfer were similar (Tables
S\ref{tab:functional-all} and S\ref{tab:go-events-recombining}).  Secondly,
either of them did not show association of virulence genes with recombining gene
transfer.  To more directly compare the methods we used the recombination
intensity and the number of recombining gene transfer events for each gene.  We
found a weak positive correlation of 0.39 with a significant P-value
of $2.2\times10^{-16}$ for testing a null hypothesis of the correlation being
zero.  (Figure S\ref{fig:cmpcomowgli}).  Interestingly, the correlation (0.03
with P-value of 0.18) disappeared in a comparison between recombination
intensity and number of horizontal gene transfer events (Figure
S\ref{fig:cmpcomowglihgt}).  

\subsection*{UCSC genome browser tracks for bacterial gene transfer study}

Often computational and statistical analyses on genomes could benefit other
researchers when those resources were provided: e.g., UCSC Genome Browser, and
Ensembl Genome Browser. We wished to provide research community especially in
SPY and SDE bacteria with some of the results from our analysis of gene
transfer. We used the UCSC genome browser for the purpose.  The resource shown
in Figure S\ref{fig:ucsc} for gene transfer is available at
\url{http://strep-genome.bscb.cornell.edu}.  For each of the five individual
genomes as a reference genome several annotation tracks were displayed:
alignment blocks, recombination rate per site for each block, local gene tree
topologies, recombination intensity, locations of virulence genes, posterior
probability of recombining gene transfer. The genome track of probability of
recombining gene transfer allowed us to easily browse the bacterial genome to
locate regions that we might be interested in. 
% To Matt: what could we show in the pan-genome? or from Mowgli's analysis?

\subsection*{Units of recombining gene transfer}

The process of gene transfer may or may not move whole genes from one species to
another \citep{Chan2009}. We observed that units of recombining gene
transfer could be multiple genes (Table S\ref{tab:genes-transfer}). In some
cases, parts of a gene were inferred to have been transferred, for instance,
transfer from SPY1 to SDE1 was inferred for 29\% of the aspartyl-tRNA synthetase
gene. Recombinant tracts were often shown to span two genes.  In other cases,
the inferred transfer units spaned a set of as much as eight adjacent genes:
loci from SpyM3\_1518 to SpyM3\_1525.

\subsection*{Simulations}

We conducted experiments on simulation data to check the accuracy of estimates
obtained by the model-based approach in a scenario similar to the one from the
real data set.  Population mutation and recombination parameters were accurately
estimated (Table S\ref{tab:sim-three}).  Yet, the recombinant tract length was
over-estimated; recombinant tract lengths were reported to be difficult to
estimate from relatively short alignment blocks \citep{Didelot2010}.  Numbers of
recombinant edges from the simulated recombinant trees of Simulation 2 (see
\textbf{Supplementary Information} for the description of Simulation 2) were
compared with the corresponding numbers of recombinant edges that were estimated
from the simulated data set from Simulation 2 (see the solid-line intervals with
open circles in Figure \ref{fig:h3}).  We noted that tree-topology-changing
recombinant edges were more reliably recovered compared with
tree-topology-not-changing ones (Figure \ref{fig:h3}). All of the
tree-topology-not-changing recombinant edge types were over-estimated (Figure
\ref{fig:h3}B). The heat map of Figure \ref{fig:Heatmap-of-recombination} did
not show how confident of the values we could be. Comparison of numbers of
recombinant edges estimated from the real data set and the corresponding numbers
estimated from Simulation 2 showed the interval estimates for the heat map
values.  Dotted-line intervals from the red and blue colored cells in the heat
map were above and below the reference of the horizontal line at the zero on the
Y-axis, respectively (Figure \ref{fig:h3}).  

For each recombinant edge type to assess the effect of posterior recombinant
trees samples with respect to that of prior recombinant tree samples. The ten
\textit{a priori} recombinant trees from Simulation 2 would have balanced
recombining gene transfer events between SPY and SDE with respect to the prior
expected number of recombinant edges. The posterior recombinant trees used in
Simulation 3 would have unbalanced recombining gene transfer between the two
species branches as we observed in the real data analysis. For significantly
strong gene flows such as ones from SPY to SDE a solid-line interval with a
closed circle would be different from its count part of the dotted-line
interval.  Figure \ref{fig:h3} showed that the recombining gene flow from SPY to
SDE was stronger enough to be nearer to the horizontal line at the zero on the
Y-axis. However, the recombining gene transfer from SDE to SPY appeared to be
too weak.  This comparison supported the biased directionality of recombining
gene transfer from SPY to SDE.

Recombination intensities that were measured on genes also showed rough
correlation (Pearson's correlation coefficient of 0.63) of the estimated and
true values (Figure S\ref{fig:ri1}).  Estimated values appeared to be driven
somewhat by the prior implied by the population parameter setting, explaining
why lower values were over-estimated and higher values were under-estimated.
Nonetheless, the correlation indicated that the measure of recombination
intensity used in our functional gene category served as an adequate indicator
of actual intensity of recombination events along genes.

